Funcoes utilizadas
Criar uma lista de comprimentos de strings em cada célula da coluna especificada
comprimentos = [len(str(celula)) for celula in df.iloc[: ,4]]
Encontrar o índice da célula com o maior número de caracteres
indice_maior = comprimentos.index(max(comprimentos))
Extrair a célula com o maior número de caracteres
celula_maior = df.iloc[indice_maior][4] df.iloc[: , 4][indice_maior] df.iloc[: , 4].value_counts() celula_maior df.iloc[:,0]
index = df.columns.get_loc(‘abcz’)
index df

IDANIMAL = CHAVE UNICA
MP = CHAVE UNICA -  Mes parto
AP = CHAVE UNICA - -Ano parto

LTOT = lactacao total
L365 = lactacao em ate 365 dias
L305 = lactacao em ate 305 dias
DE = Data de Secagem
ME = Mes de secagem
AE = ANO de Secagem
caso o LTOT, L365 e L305, DE, ME e AE sejam iguais ao arquivo old, pode ser descartado

In [2]:
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

## conectando no banco de dados

In [3]:
import mysql.connector

mydb = mysql.connector.connect(
  host="127.0.0.1",
  user="root",
  password="4d5f0l9k",
  database="bioinformatica_new"
)

## Retornando toda tabela genealogia_embrapa e genealogia_abcz e tabela criadores do banco de dados para dataframes

In [4]:
sql = f"WITH tabela_virtual AS (SELECT * FROM genealogia_embrapa) SELECT tv.*, (SELECT registro FROM tabela_virtual AS pai WHERE pai.id_produto = tv.id_touro) AS registro_pai, (SELECT registro FROM tabela_virtual AS mae WHERE mae.id_produto = tv.id_vaca) AS registro_mae FROM tabela_virtual AS tv"
sql1 = "SELECT * FROM genealogia_abcz"
sql2 = "SELECT * FROM criadores"
sql3 = "SELECT * FROM producao_embrapa"
df_genealogia_embrapa = pd.read_sql(sql, mydb)
df_genealogia_embrapa=df_genealogia_embrapa.rename(columns={'idanimal': 'IDANIMAL'})
df_genealogia_abcz = pd.read_sql(sql1, mydb)
df_criadores = pd.read_sql(sql2, mydb)
df_producao = pd.read_sql(sql3, mydb)


## Lendo planilhas

In [5]:
pedigree = pd.read_csv("pedrigree - Girolando - 17-02-2023.txt", encoding='ISO-8859-1')
dados_2023 = pd.read_csv("PNMGL - Dados girolando - 17-02-23.txt", encoding='ISO-8859-1')
dados_2022 = pd.read_csv("PNMGL - Dados_old - 17-02-2023.txt", encoding='ISO-8859-1')

In [5]:
pedigree

NOMEVACA     RGVACA  COMPRAC   IDANIMAL       DNASC SEXO  \
1                     NaN    RF-0618       11   616612.0           0    F   
2                     NaN    RF-1285       11   616611.0           0    F   
3           4527  DE IRAI    AJ-4527       34   766580.0  20/01/2006    F   
4           4530  DE IRAI    AJ-4530       34   766583.0  03/11/2004    F   
5               C551 OSKA    7932-BV       12  3410926.0  08/11/2020    F   
...                   ...        ...      ...        ...         ...  ...   
2276760           ZUZU ZU       0050       34   672063.0  29/03/2001    F   
2276761    ZUZUCA FABIANO     L-5947       34   335253.0  04/05/1995    F   
2276762  ZUZUCA FIV KENYO    0202-BI       12  2358145.0  18/02/2019    F   
2276763   ZYCH FABIANA 18  BR1235804       10  3114806.0           0    F   
2276764    ZYCH PRIMA 893  BR1537015       10  1127477.0  25/02/2009    F   

                        NOMEPAI     RGPAI  GSPAI      IDPAI  \
1                           NaN       NaN    NaN        NaN   
2                           NaN       NaN    NaN        NaN   
3                           NaN       NaN    NaN        NaN   
4                           NaN       NaN    NaN        NaN   
5                           NaN       NaN    NaN        NaN   
...                         ...       ...    ...        ...   
2276760        MIKLIN GALAXY-ET  AX102336   10.0     4653.0   
2276761                     NaN       NaN    NaN        NaN   
2276762  BACON-HILL MONTROSS-ET  AX145128   10.0  1603021.0   
2276763                     NaN       NaN    NaN        NaN   
2276764            DELTA LIONEL  AX117628   10.0   709946.0   

                    NOMEMAE      RGMAE  GSMAE      IDMAE  gs         cgen_a  \
1                       NaN        NaN    NaN        NaN   0            GIR   
2                       NaN        NaN    NaN        NaN   0            GIR   
3                       NaN        NaN    NaN        NaN   6  3/4HOL 1/4GIR   
4                       NaN        NaN    NaN        NaN   6  3/4HOL 1/4GIR   
5                       NaN        NaN    NaN        NaN   4  1/2HOL 1/2GIR   
...                     ...        ...    ...        ...  ..            ...   
2276760          AMORA Z.U.     M-8808   12.0   155290.0   6  3/4HOL 1/4GIR   
2276761                 NaN        NaN    NaN        NaN   6  3/4HOL 1/4GIR   
2276762  KATARINA FIV KENYO     KOK690   11.0  1995568.0   4  1/2HOL 1/2GIR   
2276763                 NaN        NaN    NaN        NaN   8            HOL   
2276764       ZYCH PRIMA 80  BR1369376   10.0  1318919.0   8            HOL   

         comb  
1          11  
2          11  
3         795  
4         795  
5         519  
...       ...  
2276760   795  
2276761   795  
2276762   519  
2276763   999  
2276764   999  

[2276764 rows x 17 columns]

In [6]:
dados_2022

ordem  reb_giro                           nomea  COMPRAC   RGVACA  \
0         1      2381  TULIPA MODELO JM NOVO HORIZONT       14   3670-U   
1         2      3093            ALTEROSA SM FIV MITU       12   9827-A   
2         3      2384  RODOVIA 6013 FAMOSO JM MONTE A       38   1094-T   
3         4      7453        ARAPONGA 803 FA DA SERRA       34  8590-AB   
4         5      5569   FLORA TEATRO DO SEGREDO VELHO       14   6986-J   
...     ...       ...                             ...      ...      ...   
5205   5206      9734        TIEMI CONDESSA ELUDE FIV       12  1186-AR   
5206   5207      9257         GUISA FIV MCCUTCHEN FBE       12  8181-AS   
5207   5208      9692     DAGUA MOGUL FIV MOURA LEITE       34  4050-AV   
5208   5209      5192  KARINA FIV BEEMER 2244 RS DO R       34  5851-AX   
5209   5210     11219                         6188 FQ       34  6188-AX   

                            nomep  GSPAI     RGPAI  \
0           MODELO TE DE BRASILIA     11     B5213   
1         MORNINGVIEW MISCHIEF-ET     10  AX125529   
2              FAMOSO TE SILVANIA     11    EFC779   
3           PALMCREST MONUMENT-ET     10  AX127702   
4              TEATRO DA SILVANIA     11    EFC383   
...                           ...    ...       ...   
5205          SILVERRIDGE V ELUDE     10  AX145140   
5206  DE-SU BKM MCCUTCHEN 1174-ET     10  AX140261   
5207  MOUNTFIELD SSI DCY MOGUL-ET     10  AX138953   
5208       POL BUTTE MC BEEMER-ET     10  AX145127   
5209                            0      0         0   

                               nomem  GSMAE  ... QTD_ORDENHAS        nasc  \
0     FABULA SEGREDO JM NOVO HORIZON     12  ...            2  2011-07-13   
1                   CAIANA FIV KENYO     11  ...            2  2009-07-13   
2     ATICA PARAMOUNT FIV JM MONTE A     34  ...            2  2013-03-18   
3                ADRIANA FA DA SERRA     12  ...            2  2013-10-21   
4               LUA DO SEGREDO VELHO     12  ...            2  2012-07-20   
...                              ...    ...  ...          ...         ...   
5205                   COLONIA TE AG     11  ...            2  2016-07-10   
5206                  MADAME ISIS 2B     11  ...            2  2016-08-22   
5207          QUITANDA FIV PLANET VR     12  ...            2  2017-06-22   
5208        ENGENHO DA RAINHA BIANCA     12  ...            3  2018-02-04   
5209                               0      0  ...            2  2016-04-16   

           parto     secagem anopc  anop_menos1c   rebc  rac  nordc  \
0     2018-04-26  2018-11-23  2018          2017   2381    2      2   
1     2019-03-15  2020-06-22  2019          2018   3093    2      2   
2     2019-03-26  2020-01-10  2019          2018   2384    2      2   
3     2017-10-01  2018-12-20  2017          2016   7453    2      2   
4     2020-02-01  2020-06-12  2020          2019   5569    2      2   
...          ...         ...   ...           ...    ...  ...    ...   
5205  2020-08-21  2020-12-29  2020          2019   9734    2      2   
5206  2020-05-25  2021-02-24  2020          2019   9257    3      2   
5207  2020-12-20  2021-12-05  2020          2019   9692    2      2   
5208  2020-04-01  2020-08-06  2020          2019   5192    3      3   
5209  2018-12-23  2019-12-20  2018          2017  11219    2      2   

           rebanoe  
0      23812018222  
1      30932018122  
2      23842018122  
3      74532017122  
4      55692019122  
...            ...  
5205   97342020222  
5206   92572020232  
5207   96922020122  
5208   51922020233  
5209  112192018122  

[5210 rows x 44 columns]

In [7]:
dados_2023

reb_giro                            nomea  COMPRAC   RGVACA  \
139         3093             ALTEROSA SM FIV MITU       12   9827-A   
162         9257   ESCOCIA FIV GABINETE DO R.DA P       12  8494-BA   
221         7461   NANA IMPERADOR BAXTER RF SAO F     5858  6353-AH   
222         5569    FLORA TEATRO DO SEGREDO VELHO       14   6986-J   
250        10639  LANA‡A FIV AFTERSHOCK BON.BELEM       12  9592-AG   
...          ...                              ...      ...      ...   
391297      1129            BATALHA UFP ANGICO II       38   D-0636   
391442     11844   MONTANA 1238 TEATRO FIV JM MON       12  6848-AY   
391489      6915          RAIZ FIV DAS ARABIAS II       14  3835-AD   
391670      7461   NINA IMPERADOR BAXTER RF SAO F     5858  6352-AH   
391749     83436                    LINDA SJ LALU       34  2734-BK   

                                nomep  GSPAI     RGPAI  \
139           MORNINGVIEW MISCHIEF-ET     10  AX125529   
162                 GABINETE SILVANIA     11    EFC946   
221       IMPERADOR BAXTER VOLTA FRIA     58      1459   
222                TEATRO DA SILVANIA     11    EFC383   
250     MS ATLEES SHT AFTERSHOCK - ET     10  AX129928   
...                               ...    ...       ...   
391297          DESEJO TE DA SILVANIA     11    EFC645   
391442             TEATRO DA SILVANIA     11    EFC383   
391489                   C.A.SANSAƒO      11    KCA472   
391670    IMPERADOR BAXTER VOLTA FRIA     58      1459   
391749                              0      0         0   

                                 nomem  GSMAE     RGMAE  ...  QTD_ORDENHAS  \
139                   CAIANA FIV KENYO     11     KOK31  ...             2   
162     BRAUNAS SPIRTE ROSE KATRINA-TE     10  BX381125  ...             2   
221      JURITI HERBY RF SAO FRANCISCO     58    6877-N  ...             2   
222               LUA DO SEGREDO VELHO     12    Y-5738  ...             2   
250                   JOANE FIV KUBERA     11  ACFG1840  ...             2   
...                                ...    ...       ...  ...           ...   
391297                 MIUDA ANGICO II     34   AJ-7860  ...             2   
391442            547 JM MONTE ALVERNE     10    C-0434  ...             2   
391489            PALAFITA DAS ARABIAS     12   AD-0906  ...             2   
391670  JORDANA MORRELL RF SAO FRANCIS     58    6872-N  ...             2   
391749                               0      0         0  ...             2   

              nasc       parto     secagem  anopc  anop_menos1c   rebc  rac  \
139     2009-07-13  2019-03-15  2020-06-22   2019          2018   3093    2   
162     2017-07-05  2019-12-20  2020-11-26   2019          2018   9257    2   
221     2015-03-11  2020-01-13  2021-11-11   2020          2019   7461    2   
222     2012-07-20  2020-02-01  2020-06-12   2020          2019   5569    2   
250     2014-12-22  2022-02-08  2022-10-15   2022          2021  10639    2   
...            ...         ...         ...    ...           ...    ...  ...   
391297  2008-08-25  2019-07-25  2020-04-01   2019          2018   1129    2   
391442  2016-07-14  2019-03-04  2019-06-15   2019          2018  11844    2   
391489  2011-10-29  2020-08-14  2020-10-31   2020          2019   6915    2   
391670  2015-02-23  2019-03-16  2019-12-12   2019          2018   7461    2   
391749  2017-02-04  2021-08-01  2022-04-22   2021          2020  83436    3   

        nordc       rebanoe  
139         2   30932018122  
162         2   92572019122  
221         2   74612019122  
222         2   55692019122  
250         2  106392021122  
...       ...           ...  
391297      2   11292019222  
391442      2  118442018122  
391489      2   69152020222  
391670      2   74612018122  
391749      2  834362021232  

[3872 rows x 43 columns]

In [8]:
df_genealogia_embrapa

id_produto         id  abcz registro      nome sexo paternidade  \
0            0001       0001  None     None       nan    M        None   
1            0003       0003  None     None       nan    M        None   
2            0006       0006  None     None       nan    M        None   
3       02418-249  02418-249  None     2418       nan    F        None   
4              03         03  None     None       nan    M        None   
...           ...        ...   ...      ...       ...  ...         ...   
463543       ZZZ7       ZZZ7     S     ZZZ7   AMBICAO    F        None   
463544       ZZZ8       ZZZ8     S     ZZZ8  AGASALHO    M        None   
463545       ZZZ9       ZZZ9     S     ZZZ9   BACABAL    M        None   
463546      ZZZA1      ZZZA1     S    ZZZA1     ANITA    F        None   
463547      ZZZA2      ZZZA2     S    ZZZA2    ANDREA    F        None   

       ck_sire id_touro  id_vaca  ... gr_tp     rebc2_cri      rebc2_pr   cat  \
0         None     None     None  ...     0         99999         99999  None   
1         None     None     None  ...     0         99999         99999  None   
2         None     None     None  ...     0         99999         99999  None   
3         None   108115   108615  ...     0         99999         99999  None   
4         None     None     None  ...     0         99999         99999  None   
...        ...      ...      ...  ...   ...           ...           ...   ...   
463543    None   KCA472   135177  ...     0   7145-ABCZ 1  11958-ABCZ 1    PO   
463544    None   121118   131985  ...     0   7145-ABCZ 1   7145-ABCZ 1    PO   
463545    None   133175   178922  ...     0   7145-ABCZ 1   7145-ABCZ 1    PO   
463546    None   EFC383  JDRB874  ...     0   7145-ABCZ 1   7145-ABCZ 1    PO   
463547    None  RRP5224   177993  ...     0   7145-ABCZ 1   7145-ABCZ 1    PO   

        ficha_a  cod_prod  IDANIMAL  origem  registro_pai registro_mae  
0             0         0         0  OUTRAS          None         None  
1             0         0         0  OUTRAS          None         None  
2             0         0         0  OUTRAS          None         None  
3             0         0         0  OUTRAS         A1343        18935  
4             0         0         0  OUTRAS          None         None  
...         ...       ...       ...     ...           ...          ...  
463543  1351780   3158020         0  ABCGIL        KCA472        KB429  
463544        0   3247323         0    ABCZ          K830       KA7368  
463545        0   3247322         0    ABCZ          K974       KA9422  
463546        0   9499094         0    ABCZ        EFC383      JDRB874  
463547        0   9499095         0    ABCZ       RRP5224       AB7320  

[463548 rows x 28 columns]

In [9]:
df_genealogia_abcz

cod_prod idGenealogiaABCZ    rgd   rgn sexo cat   cod_pai   cod_mae  \
0             44             None  U3983   379    F  PO    861789    849556   
1             99             None  T2799   153    F  PO   1955144    848578   
2            100             None  A7488  3755    M  PO    377108   1615593   
3            101             None  X6732  3792    F  PO   1615565   1615684   
4            102             None  X6733  3798    F  PO   1615802   1615704   
...          ...              ...    ...   ...  ...  ..       ...       ...   
745022  19650743             None   ABSA    21    F  PO  16900821  17558879   
745023  19652061             None   RRFZ    35    F  PO   9265878  15828344   
745024  19652062             None   RRFZ    37    F  PO   9265878  15186511   
745025  19652084             None   RRFZ    38    F  PO   9265878  17598634   
745026  19657679             None    EFC  2482    M  PO   9265878  13512296   

              nasc cgen_a  ... codcri  fazcri    rebc_cri     rebc2_cri  \
0       1981-10-11    GIR  ...   2761       2   2761-ABCZ   2761-ABCZ 2   
1       1977-06-05    GIR  ...    494       1    494-ABCZ    494-ABCZ 1   
2       1987-06-04    GIR  ...   2851       1   2851-ABCZ   2851-ABCZ 1   
3       1987-08-04    GIR  ...   2851       1   2851-ABCZ   2851-ABCZ 1   
4       1987-08-20    GIR  ...   2851       1   2851-ABCZ   2851-ABCZ 1   
...            ...    ...  ...    ...     ...         ...           ...   
745022  2022-07-21    GIR  ...  12626       1  12626-ABCZ  12626-ABCZ 1   
745023  2022-03-02    GIR  ...  38042       1  38042-ABCZ  38042-ABCZ 1   
745024  2022-04-29    GIR  ...  38042       1  38042-ABCZ  38042-ABCZ 1   
745025  2022-05-23    GIR  ...  38042       1  38042-ABCZ  38042-ABCZ 1   
745026  2022-07-11    GIR  ...   6949       1   6949-ABCZ   6949-ABCZ 1   

       codprop  fazprop   rebc_prop    rebc2_prop     rgd2 dt_inser  
0         7192        1   7192-ABCZ   7192-ABCZ 1    U3983     None  
1         7272        1   7272-ABCZ   7272-ABCZ 1    T2799     None  
2         7275        1   7275-ABCZ   7275-ABCZ 1    A7488     None  
3         2955        1   2955-ABCZ   2955-ABCZ 1    X6732     None  
4         7275        1   7275-ABCZ   7275-ABCZ 1    X6733     None  
...        ...      ...         ...           ...      ...      ...  
745022   12626        1  12626-ABCZ  12626-ABCZ 1   ABSA21     None  
745023   38042        1  38042-ABCZ  38042-ABCZ 1   RRFZ35     None  
745024   38042        1  38042-ABCZ  38042-ABCZ 1   RRFZ37     None  
745025   38042        1  38042-ABCZ  38042-ABCZ 1   RRFZ38     None  
745026    6949        1   6949-ABCZ   6949-ABCZ 1  EFC2482     None  

[745027 rows x 21 columns]

In [10]:
df_criadores

rebc_or idCriador                    nome_criador   codz  codgil  \
0        10-ABCZ      None            ADIR DO CARMO LEONEL     10   10010   
1      1001-GIRO      None       LEONARDO FAUSTINO MARQUES      0       0   
2           1002      None              EDUARDO F FERREIRA      0    1002   
3     10022-ABCZ      None         PALESTINA AGROPAST LTDA  10022       0   
4     10024-ABCZ      None  ANTONIO CARLOS BARBOSA ALVARES  10024       0   
...          ...       ...                             ...    ...     ...   
2367   9964-GIRO      None                             nan      0       0   
2368   9965-ABCZ      None                PAOLA GAZZINELLI   9965       0   
2369   9988-GIRO      None                             nan      0       0   
2370   9991-ABCZ      None   AGRO PASTORIL DOS POCOES LTDA   9991     272   
2371       99999      None                    DESCONHECIDO      0       0   

      codgiro            fonte data_fonte         reb  
0           0              nan       None     10-ABCZ  
1        1001  GIROLANDO202002       None   1001-GIRO  
2           0              nan       None        1002  
3           0       ABCZ202003       None  10022-ABCZ  
4           0       ABCZ202003       None  10024-ABCZ  
...       ...              ...        ...         ...  
2367     9964     GIRO20170301       None   9964-GIRO  
2368        0       ABCZ202003       None   9965-ABCZ  
2369     9988     GIRO20170301       None   9988-GIRO  
2370        0              nan       None   9991-ABCZ  
2371        0              nan       None       99999  

[2372 rows x 9 columns]

In [11]:
df_producao

chaveprod            fonte  prio_fonte  codgil   codz  faz  \
0       02418-249_1975-05             2011          36   34962   1180   10   
1           1-249_1990-12             2011          36   34962   1180   10   
2           1-281_1988-09             2011          36     281      0    0   
3           1-281_1990-05             2011          36     281      0    0   
4           1-353_1996-03             2011          36     353      0    0   
...                   ...              ...         ...     ...    ...  ...   
176707      ZZZ14_2009-05  ABCZ202112_leit           5   37040  27040   10   
176708       ZZZ7_2005-11  ABCZ202112_gord           5   21958  11958   10   
176709       ZZZ7_2007-06  ABCZ202112_leit           5   21958  11958   10   
176710       ZZZ7_2009-04  ABCZ202112_leit           5   21958  11958   10   
176711       ZZZ7_2012-04  ABCZ202112_leit           5   21958  11958   10   

        codgiro     rebc_or      rebc2_or         reb  ... psolid psol305 ccs  \
0             0  36234-ABCZ  36234-ABCZ 1  36234-ABCZ  ...      0       0   0   
1             0  36234-ABCZ  36234-ABCZ 1  36234-ABCZ  ...      0       0   0   
2             0         281           281         281  ...      0       0   0   
3             0         281           281         281  ...      0       0   0   
4             0         353           353         353  ...      0       0   0   
...         ...         ...           ...         ...  ...    ...     ...  ..   
176707        0  27040-ABCZ  27040-ABCZ 1  27040-ABCZ  ...      0       0   0   
176708        0  11958-ABCZ  11958-ABCZ 1  11958-ABCZ  ...      0       0   0   
176709        0  11958-ABCZ  11958-ABCZ 1  11958-ABCZ  ...      0       0   0   
176710        0  11958-ABCZ  11958-ABCZ 1  11958-ABCZ  ...      0       0   0   
176711        0  11958-ABCZ  11958-ABCZ 1  11958-ABCZ  ...      0       0   0   

       porgor porprot porlac  porsol  torneio pl305tor  nctor  
0        0.00     0.0    0.0     0.0     None        0      0  
1        4.39     0.0    0.0     0.0     None        0      0  
2        0.00     0.0    0.0     0.0     None        0      0  
3        3.58     0.0    0.0     0.0     None        0      0  
4        4.48     0.0    0.0     0.0     None        0      0  
...       ...     ...    ...     ...      ...      ...    ...  
176707   0.00     0.0    0.0     0.0     None        0      0  
176708   5.47     0.0    0.0     0.0     None        0      0  
176709   0.00     0.0    0.0     0.0     None        0      0  
176710   0.00     0.0    0.0     0.0     None        0      0  
176711   0.00     0.0    0.0     0.0     None        0      0  

[176712 rows x 39 columns]

## Criando um dataframe que é a junção dos dois dataframes, dados_2022 e dados_2023, pelas celulas que possuem os mesmos valores nas colunas IDANIMAL, AP, MP, L305, LTOT, L365, ME e AE.

In [6]:
df_dados_sem_atualizacao_entre_2022_2023 = pd.merge(dados_2023, dados_2022, on=['IDANIMAL', 'AP', 'MP','L305', 'LTOT', 'L365', 'ME', 'AE', 'CAUSA'],suffixes=('', '_drop'))
df_dados_sem_atualizacao_entre_2022_2023 = df_dados_sem_atualizacao_entre_2022_2023.drop(df_dados_sem_atualizacao_entre_2022_2023.filter(regex='_drop$').columns, axis=1)
df_dados_sem_atualizacao_entre_2022_2023 = df_dados_sem_atualizacao_entre_2022_2023.drop_duplicates()
df_dados_sem_atualizacao_entre_2022_2023

reb_giro                           nomea  COMPRAC   RGVACA  \
0         3093            ALTEROSA SM FIV MITU       12   9827-A   
1         9257  ESCOCIA FIV GABINETE DO R.DA P       12  8494-BA   
2         5569   FLORA TEATRO DO SEGREDO VELHO       14   6986-J   
3         7789        CABALA BRIAR 613 FUNCHAL       12  8178-AD   
4         8821  9057 WACLAW VALE OURO SILVANIA       12  4230-AI   
...        ...                             ...      ...      ...   
2828      6329             LARANJA SHOTTLE STT       12  2447-AX   
2829      1129           BATALHA UFP ANGICO II       38   D-0636   
2830     11844  MONTANA 1238 TEATRO FIV JM MON       12  6848-AY   
2831      6915         RAIZ FIV DAS ARABIAS II       14  3835-AD   
2832      7461  NINA IMPERADOR BAXTER RF SAO F     5858  6352-AH   

                            nomep  GSPAI     RGPAI  \
0         MORNINGVIEW MISCHIEF-ET     10  AX125529   
1               GABINETE SILVANIA     11    EFC946   
2              TEATRO DA SILVANIA     11    EFC383   
3             OLSZEWSKI ALTABRIAR     10  AX137375   
4           VALE OURO TE SILVANIA     11    EFC464   
...                           ...    ...       ...   
2828       MENGE SHOTTLE X-630-TE     10  AX132404   
2829        DESEJO TE DA SILVANIA     11    EFC645   
2830           TEATRO DA SILVANIA     11    EFC383   
2831                 C.A.SANSAƒO      11    KCA472   
2832  IMPERADOR BAXTER VOLTA FRIA     58      1459   

                               nomem  GSMAE     RGMAE  ...        nasc  \
0                   CAIANA FIV KENYO     11     KOK31  ...  2009-07-13   
1     BRAUNAS SPIRTE ROSE KATRINA-TE     10  BX381125  ...  2017-07-05   
2               LUA DO SEGREDO VELHO     12    Y-5738  ...  2012-07-20   
3               1760 FUNCHAL CURVELO     11    B-1760  ...  2014-10-26   
4               AGRINDUS 3432 JUNEAU     10  BX412129  ...  2015-08-10   
...                              ...    ...       ...  ...         ...   
2828                   BIRIBA TE ZBR     11     ZBR24  ...  2017-08-22   
2829                 MIUDA ANGICO II     34   AJ-7860  ...  2008-08-25   
2830            547 JM MONTE ALVERNE     10    C-0434  ...  2016-07-14   
2831            PALAFITA DAS ARABIAS     12   AD-0906  ...  2011-10-29   
2832  JORDANA MORRELL RF SAO FRANCIS     58    6872-N  ...  2015-02-23   

           parto     secagem anopc  anop_menos1c   rebc  rac  nordc  \
0     2019-03-15  2020-06-22  2019          2018   3093    2      2   
1     2019-12-20  2020-11-26  2019          2018   9257    2      2   
2     2020-02-01  2020-06-12  2020          2019   5569    2      2   
3     2018-11-23  2019-03-26  2018          2017   7789    2      2   
4     2018-10-20  2019-08-10  2018          2017   8821    2      3   
...          ...         ...   ...           ...    ...  ...    ...   
2828  2020-08-14  2021-04-21  2020          2019   6329    2      2   
2829  2019-07-25  2020-04-01  2019          2018   1129    2      2   
2830  2019-03-04  2019-06-15  2019          2018  11844    2      2   
2831  2020-08-14  2020-10-31  2020          2019   6915    2      2   
2832  2019-03-16  2019-12-12  2019          2018   7461    2      2   

           rebanoe  ordem  
0      30932018122      2  
1      92572019122    587  
2      55692019122      5  
3      77892018122   1352  
4      88212018123   5070  
...            ...    ...  
2828   63292020222   3010  
2829   11292019222   3580  
2830  118442018122   3586  
2831   69152020222   3592  
2832   74612018122   3601  

[2833 rows x 44 columns]

## Criando um dataframe com somente os novos dados inseridos em 2023
Esses são os dados que precisam ser inseridos no banco de dados

In [7]:
# faz o merge entre os dois dataframes, o dados com os dados repetidos
df_merge = pd.merge(dados_2023, df_dados_sem_atualizacao_entre_2022_2023, how='left', indicator=True, suffixes=('', '_drop'))

# filtra somente as linhas que estão no dataframe da esquerda
df_dados_novos_para_insercao_no_banco_de_dados = df_merge[df_merge['_merge'] == 'left_only']

# remove a coluna "_merge"
df_dados_novos_para_insercao_no_banco_de_dados.drop('_merge', axis=1, inplace=True)

df_dados_novos_para_insercao_no_banco_de_dados = df_dados_novos_para_insercao_no_banco_de_dados.drop(df_merge.filter(regex='_drop$').columns, axis=1)
df_dados_novos_para_insercao_no_banco_de_dados = df_dados_novos_para_insercao_no_banco_de_dados.drop_duplicates()
df_dados_novos_para_insercao_no_banco_de_dados

reb_giro                            nomea  COMPRAC    RGVACA  \
2         7461   NANA IMPERADOR BAXTER RF SAO F     5858   6353-AH   
4        10639  LANA‡A FIV AFTERSHOCK BON.BELEM       12   9592-AG   
5         6039   EDILAINE IMPERIO FIV 4303 JARD       12  AADA0352   
13        1945             ARMA MERIDIAN DA MBL       12   1025-AN   
14       73119   CELESTE 136 TEATRO FIV COOPERR       12   1239-BF   
...        ...                              ...      ...       ...   
3843      6029                DUNCAN FR RECREIO       14    0928-T   
3851       132        QUINTA I COLE AGM TRIUNFO       34   9243-AG   
3853      6915                  LIA DAS ARABIAS     5858   6125-BR   
3864      6029          GERMANA FEBO FR RECREIO     5858   9069-AS   
3871     83436                    LINDA SJ LALU       34   2734-BK   

                              nomep  GSPAI     RGPAI  \
2       IMPERADOR BAXTER VOLTA FRIA     58      1459   
4     MS ATLEES SHT AFTERSHOCK - ET     10  AX129928   
5            IMPERIO TE STA EDWIGES     11    RIG163   
13           SULLY HART MERIDIAN-ET     10  AX138576   
14               TEATRO DA SILVANIA     11    EFC383   
...                             ...    ...       ...   
3843           JAGUAR TE DO GAVIAƒO     11    GAV291   
3851         CO-OP TOYSTORY COLE-ET     10  AX129567   
3853              TURBO DAS ARABIAS   5858      0079   
3864                FEBO FR RECREIO     58    6300-N   
3871                              0      0         0   

                              nomem  GSMAE         RGMAE  ...        nasc  \
2     JURITI HERBY RF SAO FRANCISCO     58        6877-N  ...  2015-03-11   
4                  JOANE FIV KUBERA     11      ACFG1840  ...  2014-12-22   
5     SERTEC QATAR FABIANA PROGRESS     10  HBB/B-291392  ...  2013-10-18   
13                             FAMA     11        JNR167  ...  2015-04-27   
14       LAGOS ASTRE QUIMICA 795 TE     10      BX335536  ...  2016-02-05   
...                             ...    ...           ...  ...         ...   
3843           NILOPOLIS FR RECREIO     12       AB-0524  ...  2013-04-18   
3851                   QUINTA FIV M     12        0473-O  ...  2015-04-25   
3853             LILIKA DAS ARABIAS     58       4916-AW  ...  2019-06-13   
3864            CAMPANHA FR RECREIO     58        6180-N  ...  2016-07-16   
3871                              0      0             0  ...  2017-02-04   

           parto     secagem anopc  anop_menos1c   rebc  rac  nordc  \
2     2020-01-13  2021-11-11  2020          2019   7461    2      2   
4     2022-02-08  2022-10-15  2022          2021  10639    2      2   
5     2021-02-18  2021-12-20  2021          2020   6039    2      2   
13    2021-06-16  2022-02-24  2021          2020   1945    2      2   
14    2021-07-13  2022-05-22  2021          2020  73119    2      2   
...          ...         ...   ...           ...    ...  ...    ...   
3843  2021-04-22  2022-03-04  2021          2020   6029    2      2   
3851  2019-04-25  2019-09-14  2019          2018    132    2      2   
3853  2021-09-15  2022-06-10  2021          2020   6915    2      2   
3864  2019-12-10  2020-03-21  2019          2018   6029    2      2   
3871  2021-08-01  2022-04-22  2021          2020  83436    3      2   

           rebanoe  ordem  
2      74612019122    NaN  
4     106392021122    NaN  
5      60392020122    NaN  
13     19452021222    NaN  
14    731192021222    NaN  
...            ...    ...  
3843   60292021222    NaN  
3851    1322019222    NaN  
3853   69152021222    NaN  
3864   60292019122    NaN  
3871  834362021232    NaN  

[1039 rows x 44 columns]

# Adicionando coluna GS, oriunda do dataframe Pedigree, no dataframe df_dados_novos_para_insercao_no_banco_de_dados
inclusão necessária para localizar o animal

In [8]:
df_temporaria = pd.merge(df_dados_novos_para_insercao_no_banco_de_dados, pedigree[['IDANIMAL', 'gs', 'SEXO']], on=['IDANIMAL'], suffixes=('', '_drop'))
df_dados_novos_para_insercao_no_banco_de_dados=df_temporaria
df_dados_novos_para_insercao_no_banco_de_dados

nome_antigo = 'SEXO'
nome_novo = 'sexo'
indice_coluna = df_dados_novos_para_insercao_no_banco_de_dados.columns.get_loc(nome_antigo)
df_dados_novos_para_insercao_no_banco_de_dados.rename(columns={nome_antigo: nome_novo}, inplace=True)
df_dados_novos_para_insercao_no_banco_de_dados

#df_dados_novos_para_insercao_no_banco_de_dados = df_dados_novos_para_insercao_no_banco_de_dados.rename(columns={'SEXO': 'sexo'})

reb_giro                            nomea  COMPRAC    RGVACA  \
0         7461   NANA IMPERADOR BAXTER RF SAO F     5858   6353-AH   
1        10639  LANA‡A FIV AFTERSHOCK BON.BELEM       12   9592-AG   
2         6039   EDILAINE IMPERIO FIV 4303 JARD       12  AADA0352   
3         1945             ARMA MERIDIAN DA MBL       12   1025-AN   
4        73119   CELESTE 136 TEATRO FIV COOPERR       12   1239-BF   
...        ...                              ...      ...       ...   
1034      6029                DUNCAN FR RECREIO       14    0928-T   
1035       132        QUINTA I COLE AGM TRIUNFO       34   9243-AG   
1036      6915                  LIA DAS ARABIAS     5858   6125-BR   
1037      6029          GERMANA FEBO FR RECREIO     5858   9069-AS   
1038     83436                    LINDA SJ LALU       34   2734-BK   

                              nomep  GSPAI     RGPAI  \
0       IMPERADOR BAXTER VOLTA FRIA     58      1459   
1     MS ATLEES SHT AFTERSHOCK - ET     10  AX129928   
2            IMPERIO TE STA EDWIGES     11    RIG163   
3            SULLY HART MERIDIAN-ET     10  AX138576   
4                TEATRO DA SILVANIA     11    EFC383   
...                             ...    ...       ...   
1034           JAGUAR TE DO GAVIAƒO     11    GAV291   
1035         CO-OP TOYSTORY COLE-ET     10  AX129567   
1036              TURBO DAS ARABIAS   5858      0079   
1037                FEBO FR RECREIO     58    6300-N   
1038                              0      0         0   

                              nomem  GSMAE         RGMAE  ...     secagem  \
0     JURITI HERBY RF SAO FRANCISCO     58        6877-N  ...  2021-11-11   
1                  JOANE FIV KUBERA     11      ACFG1840  ...  2022-10-15   
2     SERTEC QATAR FABIANA PROGRESS     10  HBB/B-291392  ...  2021-12-20   
3                              FAMA     11        JNR167  ...  2022-02-24   
4        LAGOS ASTRE QUIMICA 795 TE     10      BX335536  ...  2022-05-22   
...                             ...    ...           ...  ...         ...   
1034           NILOPOLIS FR RECREIO     12       AB-0524  ...  2022-03-04   
1035                   QUINTA FIV M     12        0473-O  ...  2019-09-14   
1036             LILIKA DAS ARABIAS     58       4916-AW  ...  2022-06-10   
1037            CAMPANHA FR RECREIO     58        6180-N  ...  2020-03-21   
1038                              0      0             0  ...  2022-04-22   

      anopc  anop_menos1c   rebc  rac  nordc       rebanoe  ordem  gs  sexo  
0      2020          2019   7461    2      2   74612019122    NaN   5     F  
1      2022          2021  10639    2      2  106392021122    NaN   4     F  
2      2021          2020   6039    2      2   60392020122    NaN   4     F  
3      2021          2020   1945    2      2   19452021222    NaN   4     F  
4      2021          2020  73119    2      2  731192021222    NaN   4     F  
...     ...           ...    ...  ...    ...           ...    ...  ..   ...  
1034   2021          2020   6029    2      2   60292021222    NaN   2     F  
1035   2019          2018    132    2      2    1322019222    NaN   6     F  
1036   2021          2020   6915    2      2   69152021222    NaN   5     F  
1037   2019          2018   6029    2      2   60292019122    NaN   5     F  
1038   2021          2020  83436    3      2  834362021232    NaN   6     F  

[1039 rows x 46 columns]

# Criando subtabela com somente as linhas que possuem gs = 0 no dataframe df_dados_novos_para_insercao_no_banco_de_dados

In [9]:
sub_tabela_gs_0 = df_dados_novos_para_insercao_no_banco_de_dados[df_dados_novos_para_insercao_no_banco_de_dados['gs'] == 0]
sub_tabela_gs_0

reb_giro                   nomea  COMPRAC   RGVACA  \
53        5191           IBERICA FARDO       11   APKK18   
79        1945     BROMELIA FIV DO JRD       11   JRDG28   
81        4667          DAMA F.INDIANA       11  SNFP140   
99        4667    MEIA NOITE F.INDIANA       11   SNFN48   
184       5192  URUPIANGA FIV DE BRAS.       11  RRP8772   
...        ...                     ...      ...      ...   
993       5192     PILLAR FIV DE BRAS.       11  RRP8130   
995       4667      IRLANDA F. INDIANA       11   SNFN16   
1000      6915       DEJUA DAS ARABIAS       11   RHBR63   
1002      6915       PRATA DAS ARABIAS       11   RHBR49   
1005      2389   RELIQUIA FIV DE BRAS.       11  RRP8261   

                        nomep  GSPAI    RGPAI                    nomem  GSMAE  \
53          FARDO FIV F.MUTUM     11   MUT697                ESPANHOLA     11   
79           RADAR DOS POCOES     11    A7368                PROFANA 6     11   
81          FARDO FIV F.MUTUM     11   MUT697      SOBERANA F. INDIANA     11   
99         MAESTRO TE F.MUTUM     11   MUT214        ESTRELA F.INDIANA     11   
184      JAGUAR TE DO GAVIAƒO     11   GAV291       HONDA FIV DE BRAS.     11   
...                       ...    ...      ...                      ...    ...   
993   DIAMANTE TE DE BRASILIA     11  RRP5640  JAMARIA FIV DE BRASILIA     11   
995         FARDO FIV F.MUTUM     11   MUT697      LONDRINA F. INDIANA     11   
1000      MAJOR TE DOS POCOES     11  APPG801         DEJUA TE F.MUTUM     11   
1002    VALE OURO TE SILVANIA     11   EFC464                     JADE     11   
1005        FARDO FIV F.MUTUM     11   MUT697  JAMARIA FIV DE BRASILIA     11   

        RGMAE  ...     secagem  anopc  anop_menos1c  rebc  rac  nordc  \
53      APKK1  ...  2021-04-14   2020          2019  5191    2      2   
79      EBQG1  ...  2022-08-25   2022          2021  1945    2      2   
81      Z3236  ...  2022-01-28   2021          2020  4667    2      2   
99     SNFP87  ...  2022-07-06   2021          2020  4667    2      2   
184   RRP6268  ...  2022-07-13   2021          2020  5192    3      2   
...       ...  ...         ...    ...           ...   ...  ...    ...   
993   RRP6958  ...  2022-11-12   2022          2021  5192    3      2   
995    SNFP44  ...  2022-10-08   2022          2021  4667    2      2   
1000   MUT395  ...  2022-04-26   2021          2020  6915    2      2   
1002    RHBR4  ...  2022-03-01   2021          2020  6915    2      2   
1005  RRP6958  ...  2020-05-09   2019          2018  2389    2      2   

          rebanoe  ordem  gs  sexo  
53    51912020222    NaN   0     F  
79    19452022222    NaN   0     F  
81    46672021222    NaN   0     F  
99    46672020122    NaN   0     F  
184   51922021232    NaN   0     F  
...           ...    ...  ..   ...  
993   51922021132    NaN   0     F  
995   46672022222    NaN   0     F  
1000  69152021222    NaN   0     F  
1002  69152021222    NaN   0     F  
1005  23892019122    NaN   0     F  

[79 rows x 46 columns]

# Adicionando coluna cod_prod, oriunda do dataframe df_genealogia_abcz, na subtabela gs_0 dos animais que sao gs=0

In [10]:
df_genealogia_abcz = df_genealogia_abcz.rename(columns={'rgd2': 'RGVACA'})
df_temporaria1 = pd.merge(sub_tabela_gs_0, df_genealogia_abcz[['RGVACA', 'cod_prod', 'cod_pai', 'cod_mae']], on=['RGVACA'], how='left', suffixes=('', '_drop'))
sub_tabela_gs_0=df_temporaria1
sub_tabela_gs_0 = sub_tabela_gs_0.drop_duplicates()
sub_tabela_gs_0

reb_giro                   nomea  COMPRAC   RGVACA  \
0       5191           IBERICA FARDO       11   APKK18   
1       1945     BROMELIA FIV DO JRD       11   JRDG28   
2       4667          DAMA F.INDIANA       11  SNFP140   
3       4667    MEIA NOITE F.INDIANA       11   SNFN48   
4       5192  URUPIANGA FIV DE BRAS.       11  RRP8772   
..       ...                     ...      ...      ...   
74      5192     PILLAR FIV DE BRAS.       11  RRP8130   
75      4667      IRLANDA F. INDIANA       11   SNFN16   
76      6915       DEJUA DAS ARABIAS       11   RHBR63   
77      6915       PRATA DAS ARABIAS       11   RHBR49   
78      2389   RELIQUIA FIV DE BRAS.       11  RRP8261   

                      nomep  GSPAI    RGPAI                    nomem  GSMAE  \
0         FARDO FIV F.MUTUM     11   MUT697                ESPANHOLA     11   
1          RADAR DOS POCOES     11    A7368                PROFANA 6     11   
2         FARDO FIV F.MUTUM     11   MUT697      SOBERANA F. INDIANA     11   
3        MAESTRO TE F.MUTUM     11   MUT214        ESTRELA F.INDIANA     11   
4      JAGUAR TE DO GAVIAƒO     11   GAV291       HONDA FIV DE BRAS.     11   
..                      ...    ...      ...                      ...    ...   
74  DIAMANTE TE DE BRASILIA     11  RRP5640  JAMARIA FIV DE BRASILIA     11   
75        FARDO FIV F.MUTUM     11   MUT697      LONDRINA F. INDIANA     11   
76      MAJOR TE DOS POCOES     11  APPG801         DEJUA TE F.MUTUM     11   
77    VALE OURO TE SILVANIA     11   EFC464                     JADE     11   
78        FARDO FIV F.MUTUM     11   MUT697  JAMARIA FIV DE BRASILIA     11   

      RGMAE  ...  rebc  rac  nordc      rebanoe  ordem  gs  sexo    cod_prod  \
0     APKK1  ...  5191    2      2  51912020222    NaN   0     F  13672891.0   
1     EBQG1  ...  1945    2      2  19452022222    NaN   0     F  12005290.0   
2     Z3236  ...  4667    2      2  46672021222    NaN   0     F  12175694.0   
3    SNFP87  ...  4667    2      2  46672020122    NaN   0     F  12971292.0   
4   RRP6268  ...  5192    3      2  51922021232    NaN   0     F  17054525.0   
..      ...  ...   ...  ...    ...          ...    ...  ..   ...         ...   
74  RRP6958  ...  5192    3      2  51922021132    NaN   0     F  14820082.0   
75   SNFP44  ...  4667    2      2  46672022222    NaN   0     F  11921127.0   
76   MUT395  ...  6915    2      2  69152021222    NaN   0     F  13478123.0   
77    RHBR4  ...  6915    2      2  69152021222    NaN   0     F  11470928.0   
78  RRP6958  ...  2389    2      2  23892019122    NaN   0     F  15128073.0   

      cod_pai     cod_mae  
0   8530023.0  10614116.0  
1   1615985.0  10091814.0  
2   8530023.0  10905624.0  
3   5932826.0   9691642.0  
4   4176523.0  13961623.0  
..        ...         ...  
74  6716196.0  11600828.0  
75  8530023.0   6352036.0  
76  3196251.0   8006397.0  
77  3890526.0   9469537.0  
78  8530023.0  11600828.0  

[79 rows x 49 columns]

# Adicionar o cod_pai dos animais em que o pai tem o gspai = 11

In [13]:
sub_tabela_cod_pai_0 = df_dados_novos_para_insercao_no_banco_de_dados[df_dados_novos_para_insercao_no_banco_de_dados['gs_pai'] == 0]
df_temporaria2 = pd.merge(sub_tabela_cod_pai_0, df_genealogia_abcz[['RGVACA', 'gs_pai']], on=['RGVACA'], how='left', suffixes=('', '_drop'))
sub_tabela_cod_pai_0 = df_temporaria2
sub_tabela_cod_pai_0 = sub_tabela_cod_pai_0.drop_duplicates()
sub_tabela_cod_pai_0

KeyError: 'gs_pai'

# adicionar o cod_mae dos animais em que a mae tem o gsmae = 11

In [ ]:
sub_tabela_cod_mae_0 = df_dados_novos_para_insercao_no_banco_de_dados[df_dados_novos_para_insercao_no_banco_de_dados['cod_mae'] == 11]
df_temporaria3 = pd.merge(sub_tabela_cod_pai_0, df_genealogia_abcz[['RGVACA', 'cod_pai']], on=['RGVACA'], how='left', suffixes=('', '_drop'))
sub_tabela_cod_mae_0 = df_temporaria3
sub_tabela_cod_mae_0 = sub_tabela_cod_mae_0.drop_duplicates()
sub_tabela_cod_mae_0

# Adicionando codprod, cod_pai, cod_mae no dataframe df_dados_novos_para_insercao_no_banco_de_dados a partir do dataframe sub_tabela_gs_0

In [10]:
df_temporaria4 = pd.merge(df_dados_novos_para_insercao_no_banco_de_dados, sub_tabela_gs_0[['RGVACA', 'cod_prod', 'cod_pai', 'cod_mae' 'IDANIMAL']], on=['RGVACA', 'IDANIMAL'], how='left', suffixes=('', '_drop'))
df_dados_novos_para_insercao_no_banco_de_dados=df_temporaria2
df_dados_novos_para_insercao_no_banco_de_dados = df_dados_novos_para_insercao_no_banco_de_dados.drop_duplicates()
df_dados_novos_para_insercao_no_banco_de_dados

reb_giro                            nomea  COMPRAC    RGVACA  \
0         7461   NANA IMPERADOR BAXTER RF SAO F     5858   6353-AH   
1        10639  LANA‡A FIV AFTERSHOCK BON.BELEM       12   9592-AG   
2         6039   EDILAINE IMPERIO FIV 4303 JARD       12  AADA0352   
3         1945             ARMA MERIDIAN DA MBL       12   1025-AN   
4        73119   CELESTE 136 TEATRO FIV COOPERR       12   1239-BF   
...        ...                              ...      ...       ...   
1038      6029                DUNCAN FR RECREIO       14    0928-T   
1039       132        QUINTA I COLE AGM TRIUNFO       34   9243-AG   
1040      6915                  LIA DAS ARABIAS     5858   6125-BR   
1041      6029          GERMANA FEBO FR RECREIO     5858   9069-AS   
1042     83436                    LINDA SJ LALU       34   2734-BK   

                              nomep  GSPAI     RGPAI  \
0       IMPERADOR BAXTER VOLTA FRIA     58      1459   
1     MS ATLEES SHT AFTERSHOCK - ET     10  AX129928   
2            IMPERIO TE STA EDWIGES     11    RIG163   
3            SULLY HART MERIDIAN-ET     10  AX138576   
4                TEATRO DA SILVANIA     11    EFC383   
...                             ...    ...       ...   
1038           JAGUAR TE DO GAVIAƒO     11    GAV291   
1039         CO-OP TOYSTORY COLE-ET     10  AX129567   
1040              TURBO DAS ARABIAS   5858      0079   
1041                FEBO FR RECREIO     58    6300-N   
1042                              0      0         0   

                              nomem  GSMAE         RGMAE  ...  anopc  \
0     JURITI HERBY RF SAO FRANCISCO     58        6877-N  ...   2020   
1                  JOANE FIV KUBERA     11      ACFG1840  ...   2022   
2     SERTEC QATAR FABIANA PROGRESS     10  HBB/B-291392  ...   2021   
3                              FAMA     11        JNR167  ...   2021   
4        LAGOS ASTRE QUIMICA 795 TE     10      BX335536  ...   2021   
...                             ...    ...           ...  ...    ...   
1038           NILOPOLIS FR RECREIO     12       AB-0524  ...   2021   
1039                   QUINTA FIV M     12        0473-O  ...   2019   
1040             LILIKA DAS ARABIAS     58       4916-AW  ...   2021   
1041            CAMPANHA FR RECREIO     58        6180-N  ...   2019   
1042                              0      0             0  ...   2021   

      anop_menos1c   rebc rac  nordc       rebanoe  ordem  gs  sexo  cod_prod  
0             2019   7461   2      2   74612019122    NaN   5     F       NaN  
1             2021  10639   2      2  106392021122    NaN   4     F       NaN  
2             2020   6039   2      2   60392020122    NaN   4     F       NaN  
3             2020   1945   2      2   19452021222    NaN   4     F       NaN  
4             2020  73119   2      2  731192021222    NaN   4     F       NaN  
...            ...    ...  ..    ...           ...    ...  ..   ...       ...  
1038          2020   6029   2      2   60292021222    NaN   2     F       NaN  
1039          2018    132   2      2    1322019222    NaN   6     F       NaN  
1040          2020   6915   2      2   69152021222    NaN   5     F       NaN  
1041          2018   6029   2      2   60292019122    NaN   5     F       NaN  
1042          2020  83436   3      2  834362021232    NaN   6     F       NaN  

[1039 rows x 47 columns]

# Adicionando cod_pai no df_dados_novos oriundos da subtabela sub_tabela_cod_pai_0

In [ ]:
df_temporaria5 = df_dados_novos_para_insercao_no_banco_de_dados.merge(sub_tabela_cod_pai_0[['ID_ANIMAL', 'RGVACA', 'cod_pai']], on=['ID_ANIMAL', 'RGVACA'], how='left')
df_temporaria5


# Adicionando cod_mae no df_dados_novos oriundos da subtabela sub_tabela_cod_mae_0

In [ ]:
df_temporaria6 = pd.merge(df_dados_novos_para_insercao_no_banco_de_dados, sub_tabela_cod_pai_0[['RGVACA' 'cod_mae', 'IDANIMAL']], on=['RGVACA', 'IDANIMAL'], how='left', suffixes=('', '_drop'))
df_dados_novos_para_insercao_no_banco_de_dados=df_temporaria2
df_dados_novos_para_insercao_no_banco_de_dados = df_dados_novos_para_insercao_no_banco_de_dados.drop_duplicates()
df_dados_novos_para_insercao_no_banco_de_dados

# Adicionando reb no dataframe df_dados_novos_para_insercao_no_banco_de_dados
Obs: Cod_gil esta na tabela de criadores

In [11]:
df_criadores = df_criadores.rename(columns={'codgiro':'reb_giro'})

df_dados_novos_para_insercao_no_banco_de_dados = pd.merge(df_dados_novos_para_insercao_no_banco_de_dados, df_criadores[['reb_giro', 'reb']], on=['reb_giro'], how='left')
df_dados_novos_para_insercao_no_banco_de_dados = df_dados_novos_para_insercao_no_banco_de_dados.drop_duplicates()
df_dados_novos_para_insercao_no_banco_de_dados

reb_giro                            nomea  COMPRAC    RGVACA  \
0         7461   NANA IMPERADOR BAXTER RF SAO F     5858   6353-AH   
2        10639  LANA‡A FIV AFTERSHOCK BON.BELEM       12   9592-AG   
3         6039   EDILAINE IMPERIO FIV 4303 JARD       12  AADA0352   
4         1945             ARMA MERIDIAN DA MBL       12   1025-AN   
5        73119   CELESTE 136 TEATRO FIV COOPERR       12   1239-BF   
...        ...                              ...      ...       ...   
1150      6029                DUNCAN FR RECREIO       14    0928-T   
1151       132        QUINTA I COLE AGM TRIUNFO       34   9243-AG   
1152      6915                  LIA DAS ARABIAS     5858   6125-BR   
1153      6029          GERMANA FEBO FR RECREIO     5858   9069-AS   
1154     83436                    LINDA SJ LALU       34   2734-BK   

                              nomep  GSPAI     RGPAI  \
0       IMPERADOR BAXTER VOLTA FRIA     58      1459   
2     MS ATLEES SHT AFTERSHOCK - ET     10  AX129928   
3            IMPERIO TE STA EDWIGES     11    RIG163   
4            SULLY HART MERIDIAN-ET     10  AX138576   
5                TEATRO DA SILVANIA     11    EFC383   
...                             ...    ...       ...   
1150           JAGUAR TE DO GAVIAƒO     11    GAV291   
1151         CO-OP TOYSTORY COLE-ET     10  AX129567   
1152              TURBO DAS ARABIAS   5858      0079   
1153                FEBO FR RECREIO     58    6300-N   
1154                              0      0         0   

                              nomem  GSMAE         RGMAE  ...  anop_menos1c  \
0     JURITI HERBY RF SAO FRANCISCO     58        6877-N  ...          2019   
2                  JOANE FIV KUBERA     11      ACFG1840  ...          2021   
3     SERTEC QATAR FABIANA PROGRESS     10  HBB/B-291392  ...          2020   
4                              FAMA     11        JNR167  ...          2020   
5        LAGOS ASTRE QUIMICA 795 TE     10      BX335536  ...          2020   
...                             ...    ...           ...  ...           ...   
1150           NILOPOLIS FR RECREIO     12       AB-0524  ...          2020   
1151                   QUINTA FIV M     12        0473-O  ...          2018   
1152             LILIKA DAS ARABIAS     58       4916-AW  ...          2020   
1153            CAMPANHA FR RECREIO     58        6180-N  ...          2018   
1154                              0      0             0  ...          2020   

       rebc  rac nordc       rebanoe  ordem  gs  sexo  cod_prod         reb  
0      7461    2     2   74612019122    NaN   5     F       NaN        1291  
2     10639    2     2  106392021122    NaN   4     F       NaN  10639-GIRO  
3      6039    2     2   60392020122    NaN   4     F       NaN         NaN  
4      1945    2     2   19452021222    NaN   4     F       NaN  37634-ABCZ  
5     73119    2     2  731192021222    NaN   4     F       NaN         NaN  
...     ...  ...   ...           ...    ...  ..   ...       ...         ...  
1150   6029    2     2   60292021222    NaN   2     F       NaN  31365-ABCZ  
1151    132    2     2    1322019222    NaN   6     F       NaN  31835-ABCZ  
1152   6915    2     2   69152021222    NaN   5     F       NaN  27729-ABCZ  
1153   6029    2     2   60292019122    NaN   5     F       NaN  31365-ABCZ  
1154  83436    3     2  834362021232    NaN   6     F       NaN  83436-GIRO  

[1039 rows x 48 columns]

In [12]:
#verifica_reb = df_criadores.groupby('reb_giro')['reb'].nunique()
#verifica_reb

## Calculando a diferença entre a data dos novos partos com as ultimas datas presentes na tabela de dados_2022 para validar se ha alguma inconsistência

obs: Caso a diferença seja menor que 250 dias é provável que haja alguma falha.

In [13]:

df_diferenca_entre_partos_novos_e_do_banco = pd.merge(df_dados_novos_para_insercao_no_banco_de_dados, dados_2022, on=['IDANIMAL'], suffixes=('', '_drop'))
df_diferenca_entre_partos_novos_e_do_banco = df_diferenca_entre_partos_novos_e_do_banco.drop_duplicates()
#df_diferenca_entre_partos_novos_e_do_banco = df_diferenca_entre_partos_novos_e_do_banco.drop(df_diferenca_entre_partos_novos_e_do_banco.filter(regex='_drop$').columns, axis=1)
df_diferenca_entre_partos_novos_e_do_banco[['parto', 'parto_drop']] = df_diferenca_entre_partos_novos_e_do_banco[['parto', 'parto_drop']].astype(str)
df_diferenca_entre_partos_novos_e_do_banco[['parto', 'parto_drop']] = df_diferenca_entre_partos_novos_e_do_banco[['parto', 'parto_drop']].apply(pd.to_datetime)
df_diferenca_entre_partos_novos_e_do_banco['diferenca_em_dias'] = (df_diferenca_entre_partos_novos_e_do_banco['parto'] - df_diferenca_entre_partos_novos_e_do_banco['parto_drop']).dt.days
df_diferenca_entre_partos_novos_e_do_banco[['parto', 'parto_drop', 'diferenca_em_dias']]

parto parto_drop  diferenca_em_dias
0   2021-11-23 2020-01-09                684
1   2020-10-09 2019-01-31                617
2   2021-03-15 2020-01-24                416
3   2021-03-15 2018-02-01               1138
4   2021-06-28 2020-07-22                341
..         ...        ...                ...
682 2021-03-23 2017-11-17               1222
683 2021-03-23 2020-03-10                378
684 2021-01-25 2019-05-03                633
685 2021-04-22 2020-05-13                344
686 2021-08-01 2020-01-05                574

[687 rows x 3 columns]

## Pegando da tabela de animais que serão inseridos no banco aqueles que ja possuem IDANIMAL cadastrado
Comparado ao dataframe  dados_2022

In [14]:
df_animais_com_ID = pd.merge(df_dados_novos_para_insercao_no_banco_de_dados, dados_2022, on =['IDANIMAL'], suffixes=('', '_drop'))
df_animais_com_ID = df_animais_com_ID.drop(df_animais_com_ID.filter(regex='_drop$').columns, axis=1)
df_animais_com_ID = df_animais_com_ID.drop_duplicates()
df_animais_com_ID

reb_giro                            nomea  COMPRAC   RGVACA  \
0        6029               DEMASIA FR RECREIO       14   0869-T   
1        7450                JACA JAYZ SJ LALU       34  6411-AH   
2        3304   CAITA OCIDENTE MJP DA CAPELA P       12   8634-X   
4        5569   LONDRINA GENGIS KHAN RECANTO D       12  0021-AW   
5        5569          BABALU METEORO DO MORRO       38  0148-AK   
..        ...                              ...      ...      ...   
680      1129                   MORADIA ANGICO        7   9556-F   
681      2381  CLASSICA OCIDENTE JM NOVO HORI       14  5193-AE   
684      5569           GABI FIV VALE OURO EVD       14  4417-AT   
685      6029                DUNCAN FR RECREIO       14   0928-T   
686     83436                    LINDA SJ LALU       34  2734-BK   

                       nomep  GSPAI     RGPAI                           nomem  \
0       JAGUAR TE DO GAVIAƒO     11    GAV291            ERMELINDA MERGULHAO    
1           HORNLAND JAYZ-ET     10  AX118150         GAYA FIV SANSAO SJ LALU   
2    OCIDENTE BEM FEITOR CAL     11   CAL4605          MARINA DA CAPELA PAIVA   
4    GENGIS KHAN DE BRASILIA     11   RRP6097   MEARA JUDITH GOLDWYN BELLA-TE   
5        METEORO DE BRASILIA     11     B5226                SEGUNDA DO MORRO   
..                       ...    ...       ...                             ...   
680    VALE OURO DE BRASILIA     11     A6796                   CABANA ANGICO   
681  OCIDENTE BEM FEITOR CAL     11   CAL4605  CINDERELA MODELO JM NOVO HORIZ   
684    VALE OURO TE SILVANIA     11    EFC464    LAMBADA DRAMATIC SANTA LUZIA   
685     JAGUAR TE DO GAVIAƒO     11    GAV291            NILOPOLIS FR RECREIO   
686                        0      0         0                               0   

     GSMAE     RGMAE  ...  anop_menos1c   rebc  rac nordc       rebanoe  \
0       12   AD-1407  ...          2020   6029    2     2   60292021122   
1       12    1898-P  ...          2019   7450    3     2   74502020132   
2       10    A-3096  ...          2020   3304    2     2   33042020122   
4       10  BX410396  ...          2020   5569    2     2   55692021222   
5       34    7468-T  ...          2020   5569    2     2   55692021222   
..     ...       ...  ...           ...    ...  ...   ...           ...   
680     78      1606  ...          2020   1129    2     2   11292021222   
681     12   AF-2652  ...          2020   2381    2     2   23812020122   
684     12   AG-0944  ...          2020   5569    2     2   55692020122   
685     12   AB-0524  ...          2020   6029    2     2   60292021222   
686      0         0  ...          2020  83436    3     2  834362021232   

     ordem  gs  sexo  cod_prod         reb  
0      NaN   2     F       NaN  31365-ABCZ  
1      NaN   6     F       NaN   7450-GIRO  
2      NaN   4     F       NaN   3304-GIRO  
4      NaN   4     F       NaN   5569-GIRO  
5      NaN   3     F       NaN   5569-GIRO  
..     ...  ..   ...       ...         ...  
680    NaN   4     F       NaN       52721  
681    NaN   2     F       NaN  31300-ABCZ  
684    NaN   2     F       NaN   5569-GIRO  
685    NaN   2     F       NaN  31365-ABCZ  
686    NaN   6     F       NaN  83436-GIRO  

[417 rows x 48 columns]

## Pegando da tabela de animais que serão inseridos no banco, todos aqueles que não possuem IDANIMAL cadastrado
comparado ao dataframe dados_2022

In [15]:
df_animais_sem_id = pd.merge(df_dados_novos_para_insercao_no_banco_de_dados, dados_2022, on='IDANIMAL', how='left', indicator=True, suffixes=('', '_drop'))
df_animais_sem_id = df_animais_sem_id.drop(df_animais_sem_id.filter(regex='_drop$').columns, axis=1)
df_animais_sem_id = df_animais_sem_id[df_animais_sem_id['_merge'] == 'left_only'].drop(columns='_merge')
df_animais_sem_id

reb_giro                            nomea  COMPRAC    RGVACA  \
0         7461   NANA IMPERADOR BAXTER RF SAO F     5858   6353-AH   
1        10639  LANA‡A FIV AFTERSHOCK BON.BELEM       12   9592-AG   
2         6039   EDILAINE IMPERIO FIV 4303 JARD       12  AADA0352   
3         1945             ARMA MERIDIAN DA MBL       12   1025-AN   
4        73119   CELESTE 136 TEATRO FIV COOPERR       12   1239-BF   
...        ...                              ...      ...       ...   
1301      1945   JPZ QUIRERA SHAMROCK ASTER FIV       12    4032-Z   
1302      2381   DALIA I FIV BRADLEY JM NOVO HO       12   4906-AZ   
1305       132        QUINTA I COLE AGM TRIUNFO       34   9243-AG   
1306      6915                  LIA DAS ARABIAS     5858   6125-BR   
1307      6029          GERMANA FEBO FR RECREIO     5858   9069-AS   

                              nomep  GSPAI     RGPAI  \
0       IMPERADOR BAXTER VOLTA FRIA     58      1459   
1     MS ATLEES SHT AFTERSHOCK - ET     10  AX129928   
2            IMPERIO TE STA EDWIGES     11    RIG163   
3            SULLY HART MERIDIAN-ET     10  AX138576   
4                TEATRO DA SILVANIA     11    EFC383   
...                             ...    ...       ...   
1301     LADYS-MANOR PL SHAMROCK-ET     10  AX135714   
1302               ROCKALLI BRADLEY     10  AX117720   
1305         CO-OP TOYSTORY COLE-ET     10  AX129567   
1306              TURBO DAS ARABIAS   5858      0079   
1307                FEBO FR RECREIO     58    6300-N   

                              nomem  GSMAE         RGMAE  ...  anop_menos1c  \
0     JURITI HERBY RF SAO FRANCISCO     58        6877-N  ...          2019   
1                  JOANE FIV KUBERA     11      ACFG1840  ...          2021   
2     SERTEC QATAR FABIANA PROGRESS     10  HBB/B-291392  ...          2020   
3                              FAMA     11        JNR167  ...          2020   
4        LAGOS ASTRE QUIMICA 795 TE     10      BX335536  ...          2020   
...                             ...    ...           ...  ...           ...   
1301                  JPZ ASTER FIV     11         JPZ64  ...          2021   
1302           DALIA NOVO HORIZONTE     11        JMCH76  ...          2020   
1305                   QUINTA FIV M     12        0473-O  ...          2018   
1306             LILIKA DAS ARABIAS     58       4916-AW  ...          2020   
1307            CAMPANHA FR RECREIO     58        6180-N  ...          2018   

       rebc  rac nordc       rebanoe  ordem  gs  sexo  cod_prod         reb  
0      7461    2     2   74612019122    NaN   5     F       NaN        1291  
1     10639    2     2  106392021122    NaN   4     F       NaN  10639-GIRO  
2      6039    2     2   60392020122    NaN   4     F       NaN         NaN  
3      1945    2     2   19452021222    NaN   4     F       NaN  37634-ABCZ  
4     73119    2     2  731192021222    NaN   4     F       NaN         NaN  
...     ...  ...   ...           ...    ...  ..   ...       ...         ...  
1301   1945    2     2   19452022222    NaN   4     F       NaN  37634-ABCZ  
1302   2381    2     2   23812021222    NaN   4     F       NaN  31300-ABCZ  
1305    132    2     2    1322019222    NaN   6     F       NaN  31835-ABCZ  
1306   6915    2     2   69152021222    NaN   5     F       NaN  27729-ABCZ  
1307   6029    2     2   60292019122    NaN   5     F       NaN  31365-ABCZ  

[622 rows x 48 columns]

In [16]:
#dados_2022[dados_2022['IDANIMAL'] == 1631099]
#df_verificando_animais_com_id_na_genealogia_embrapa['IDANIMAL'].value_counts()
#verifica_reb = df_verificando_animais_com_id_na_genealogia_embrapa.groupby('IDANIMAL')['id_produto'].nunique()
#verifica_reb

# Adicionando a coluna id_produto ao dataframe df_animais_com_ID

In [17]:
df_genealogia_embrapa['IDANIMAL'] = df_genealogia_embrapa['IDANIMAL'].astype('int')
df_animais_com_ID = pd.merge(df_animais_com_ID, df_genealogia_embrapa[['IDANIMAL', 'id_produto']], on='IDANIMAL', how='left', indicator=True)
df_animais_com_ID = df_animais_com_ID.drop_duplicates()
df_animais_com_ID = df_animais_com_ID.drop(columns='_merge')
df_animais_com_ID

reb_giro                            nomea  COMPRAC   RGVACA  \
0        6029               DEMASIA FR RECREIO       14   0869-T   
1        7450                JACA JAYZ SJ LALU       34  6411-AH   
2        3304   CAITA OCIDENTE MJP DA CAPELA P       12   8634-X   
3        5569   LONDRINA GENGIS KHAN RECANTO D       12  0021-AW   
4        5569          BABALU METEORO DO MORRO       38  0148-AK   
..        ...                              ...      ...      ...   
415      1129                   MORADIA ANGICO        7   9556-F   
416      2381  CLASSICA OCIDENTE JM NOVO HORI       14  5193-AE   
417      5569           GABI FIV VALE OURO EVD       14  4417-AT   
418      6029                DUNCAN FR RECREIO       14   0928-T   
419     83436                    LINDA SJ LALU       34  2734-BK   

                       nomep  GSPAI     RGPAI                           nomem  \
0       JAGUAR TE DO GAVIAƒO     11    GAV291            ERMELINDA MERGULHAO    
1           HORNLAND JAYZ-ET     10  AX118150         GAYA FIV SANSAO SJ LALU   
2    OCIDENTE BEM FEITOR CAL     11   CAL4605          MARINA DA CAPELA PAIVA   
3    GENGIS KHAN DE BRASILIA     11   RRP6097   MEARA JUDITH GOLDWYN BELLA-TE   
4        METEORO DE BRASILIA     11     B5226                SEGUNDA DO MORRO   
..                       ...    ...       ...                             ...   
415    VALE OURO DE BRASILIA     11     A6796                   CABANA ANGICO   
416  OCIDENTE BEM FEITOR CAL     11   CAL4605  CINDERELA MODELO JM NOVO HORIZ   
417    VALE OURO TE SILVANIA     11    EFC464    LAMBADA DRAMATIC SANTA LUZIA   
418     JAGUAR TE DO GAVIAƒO     11    GAV291            NILOPOLIS FR RECREIO   
419                        0      0         0                               0   

     GSMAE     RGMAE  ...   rebc  rac  nordc       rebanoe  ordem  gs  sexo  \
0       12   AD-1407  ...   6029    2      2   60292021122    NaN   2     F   
1       12    1898-P  ...   7450    3      2   74502020132    NaN   6     F   
2       10    A-3096  ...   3304    2      2   33042020122    NaN   4     F   
3       10  BX410396  ...   5569    2      2   55692021222    NaN   4     F   
4       34    7468-T  ...   5569    2      2   55692021222    NaN   3     F   
..     ...       ...  ...    ...  ...    ...           ...    ...  ..   ...   
415     78      1606  ...   1129    2      2   11292021222    NaN   4     F   
416     12   AF-2652  ...   2381    2      2   23812020122    NaN   2     F   
417     12   AG-0944  ...   5569    2      2   55692020122    NaN   2     F   
418     12   AB-0524  ...   6029    2      2   60292021222    NaN   2     F   
419      0         0  ...  83436    3      2  834362021232    NaN   6     F   

     cod_prod         reb   id_produto  
0         NaN  31365-ABCZ  GIRO1383626  
1         NaN   7450-GIRO  GIRO1608903  
2         NaN   3304-GIRO  GIRO1215331  
3         NaN   5569-GIRO  GIRO2022656  
4         NaN   5569-GIRO  GIRO1653303  
..        ...         ...          ...  
415       NaN       52721  GIRO1066128  
416       NaN  31300-ABCZ  GIRO1571455  
417       NaN   5569-GIRO  GIRO1797713  
418       NaN  31365-ABCZ  GIRO1383629  
419       NaN  83436-GIRO  GIRO2900035  

[420 rows x 49 columns]

## Lista de animais do dataframe sem ID em relacao a dados_2022 e que possuem ID_animal no dataframe da genealogia_Embrapa
Obs: Tem que ter ID_produto

In [18]:
df_animais_sem_id = df_animais_sem_id.add_suffix('_esq')
df_genealogia_embrapa = df_genealogia_embrapa.add_suffix('_dir')

df_animais_sem_id_2022_que_possuem_id_no_banco = pd.merge(df_animais_sem_id, df_genealogia_embrapa, left_on=['IDANIMAL_esq'], right_on=['IDANIMAL_dir'], how='left', indicator=True)
df_animais_sem_id_2022_que_possuem_id_no_banco.drop_duplicates()
df_animais_sem_id_2022_que_possuem_id_no_banco = df_animais_sem_id_2022_que_possuem_id_no_banco.drop(df_animais_sem_id_2022_que_possuem_id_no_banco.filter(regex='_drop$').columns, axis=1)
df_animais_sem_id_2022_que_possuem_id_no_banco = df_animais_sem_id_2022_que_possuem_id_no_banco[df_animais_sem_id_2022_que_possuem_id_no_banco['_merge'] == 'both']

# dropa as colunas que não são necessárias
df_animais_sem_id_2022_que_possuem_id_no_banco = df_animais_sem_id_2022_que_possuem_id_no_banco.drop(df_animais_sem_id_2022_que_possuem_id_no_banco.filter(regex='_dir').columns.difference(['id_produto_dir']), axis=1)
df_animais_sem_id_2022_que_possuem_id_no_banco = df_animais_sem_id_2022_que_possuem_id_no_banco.rename(columns=lambda x: x.replace('_esq', '') if '_esq' in x else x)
df_animais_sem_id_2022_que_possuem_id_no_banco = df_animais_sem_id_2022_que_possuem_id_no_banco.rename(columns=lambda x: x.replace('_dir', '') if '_dir' in x else x)
df_animais_sem_id_2022_que_possuem_id_no_banco.drop(columns='_merge')
df_animais_sem_id_2022_que_possuem_id_no_banco = df_animais_sem_id_2022_que_possuem_id_no_banco.drop(columns='_merge')
df_animais_sem_id_2022_que_possuem_id_no_banco

reb_giro                           nomea  COMPRAC    RGVACA  \
9        1945  JPZ RENDEIRA TEATRO GENOVA FIV       12   7881-AD   
21       7461        PORCINA RF SAO FRANCISCO       38   4082-AX   
25       9000            ICH Q379 CANELA TABU       14   9488-AN   
45       2915          BOLIVIA SHREK 2R JATAI       12    3784-M   
46       8812  BATALHA VALE OURO SERRA DO LUA       12    3200-J   
58       7461  IRONDINA ROGER RF SAO FRANCISC       58    6886-N   
70       7450    JEITOSA FIV DELEGADO SJ LALU       12   6403-AH   
98       7461          ELITE RF SAO FRANCISCO       78    6715-N   
136      5426  HEGLA BELLA ESPELHO FIV DA ZEN       12   1080-AM   
161      4237          CARLA EXPOENTE FUNDAƒO       14    7344-U   
184      5678            GBEZ 3016 GOMEATINGA       34    9191-T   
188      2915    JPZ PANCAS TEATRO GENOVA FIV       12    3181-Z   
210      8030  CHACACA SANSAO FIV DO FAZENDAO       14    B-7822   
229      5551  INVEJOSA SANSAO FIV TS DA MUQU       14    7661-Z   
293      7461          MOELA RF SAO FRANCISCO       38   6330-AH   
298      3717           IATINA TABU FIV DELIB       14    8003-N   
311      8030             RENDA BLITZ FIV FZD       12    8671-I   
344      5678             GBEZ3004 GOMEATINGA       38   6948-AB   
369      5191          HAVAIANA JAGUAR FIV 2K       12    0534-Q   
388      6029            CAJAZEIRA FR RECREIO       14    6177-N   
389      6029            CAJAZEIRA FR RECREIO       14    6177-N   
393      5497               HELLEN VAIDOSO SV       14    0112-X   
395     10639           DELICADA DO MEU SONHO       11    GGMS59   
414      7461         OLENDA RF SAO FRANCISCO       38   2683-AS   
432      5551     BB MILK  VICKY-UP FACHO FIV       14   5639-AG   
443     11818        ENCANTADA FIV DA CRAUATA       11     AOI54   
446      7754  IMPONENCIA FIV SANSAO SANTA LU       14    7393-Z   
456      1026    LALICA VALE OURO FBS CAETANO       12    4442-Y   
469      5192  HIPEREIDES CASTELO FIV EMILY R       12   6284-AL   
487      5191                JUARA GALEAƒO 2K       12   6887-AH   
488      5191                        GRACIOSA       11     APKK9   
492      5407  GALOCHA JAGUAR FIV DA CACHOEIR       14   6545-AL   
499      2915                        CARGA IA       14   0079-AX   
500      1129                 JORDANIA ANGICO        7      0484   
535      8689         JOA EMBAIXADOR LAGLORIA       14    6341-M   
544      9373            JAB BARONEZA OPINION       78   8900-AI   
563      5192          SUGESTIVA FIV DE BRAS.       11   RRP8637   
569      6029                GRUTA FR RECREIO       12   4847-BG   
591      1945           ABAMA 2 FIV M. VERDE        11  ISPG1286   

                       nomep  GSPAI     RGPAI  \
9         TEATRO DA SILVANIA     11    EFC383   
21         EURO TE JABAQUARA     11   EVPF425   
25            TABU TE DA CAL     11   CAL6557   
45          J.D.F. SHREK 932     10  AX115283   
46     VALE OURO TE SILVANIA     11    EFC464   
58         FREUREHAVEN ROGER     10  AX125592   
70                 DELEGADO      11   BJAS204   
98                         0      0         0   
136   ESPELHO TE DE BRASILIA     11   RRP5664   
161     EXPOENTE DE BRASILIA     11    B-0208   
184    PALMCREST MONUMENT-ET     10  AX127702   
188       TEATRO DA SILVANIA     11    EFC383   
210             C.A.SANSAƒO      11    KCA472   
229             C.A.SANSAƒO      11    KCA472   
293            BARAƒO DO JRD     11    JRDG15   
298           TABU TE DA CAL     11   CAL6557   
311   FUSTEAD EMORY BLITZ-ET     10  AX111249   
344     JAGUAR TE DO GAVIAƒO     11    GAV291   
369     JAGUAR TE DO GAVIAƒO     11    GAV291   
388  OCIDENTE BEM FEITOR CAL     11   CAL4605   
389  OCIDENTE BEM FEITOR CAL     11   CAL4605   
393      VAIDOSO DA SILVANIA     11    EFC441   
395         RADAR DOS POCOES     11     A7368   
414        GARANTIDO DA JGVA     11   JGVA252   
432          FACHO TE KUBERA     11   ACFG834   
443  DIAMA

# Criando dataframe com todos os animais que possuem ID encontrado

In [19]:
# Juntando os dataframes
df_animais_com_ID = pd.concat([df_animais_com_ID, df_animais_sem_id_2022_que_possuem_id_no_banco])
df_animais_com_ID

reb_giro                           nomea  COMPRAC    RGVACA  \
0        6029              DEMASIA FR RECREIO       14    0869-T   
1        7450               JACA JAYZ SJ LALU       34   6411-AH   
2        3304  CAITA OCIDENTE MJP DA CAPELA P       12    8634-X   
3        5569  LONDRINA GENGIS KHAN RECANTO D       12   0021-AW   
4        5569         BABALU METEORO DO MORRO       38   0148-AK   
..        ...                             ...      ...       ...   
535      8689         JOA EMBAIXADOR LAGLORIA       14    6341-M   
544      9373            JAB BARONEZA OPINION       78   8900-AI   
563      5192          SUGESTIVA FIV DE BRAS.       11   RRP8637   
569      6029                GRUTA FR RECREIO       12   4847-BG   
591      1945           ABAMA 2 FIV M. VERDE        11  ISPG1286   

                       nomep  GSPAI     RGPAI                          nomem  \
0       JAGUAR TE DO GAVIAƒO     11    GAV291           ERMELINDA MERGULHAO    
1           HORNLAND JAYZ-ET     10  AX118150        GAYA FIV SANSAO SJ LALU   
2    OCIDENTE BEM FEITOR CAL     11   CAL4605         MARINA DA CAPELA PAIVA   
3    GENGIS KHAN DE BRASILIA     11   RRP6097  MEARA JUDITH GOLDWYN BELLA-TE   
4        METEORO DE BRASILIA     11     B5226               SEGUNDA DO MORRO   
..                       ...    ...       ...                            ...   
535      EMBAIXADOR LAGLORIA     11    MAFV20              FLORINDA LAGLORIA   
544  ROYAL OPINION DECEMBER      10  AX119971  INTEGRALAT FIV 77761 KIRKLAND   
563     JAGUAR TE DO GAVIAƒO     11    GAV291               JESSICA DE BRAS.   
569                        0      0         0                              0   
591             C.A.SANSAƒO      11    KCA472             ABAMA TE DE KUBERA   

     GSMAE     RGMAE  ...  rebc  rac  nordc      rebanoe  ordem  gs  sexo  \
0       12   AD-1407  ...  6029    2      2  60292021122    NaN   2     F   
1       12    1898-P  ...  7450    3      2  74502020132    NaN   6     F   
2       10    A-3096  ...  3304    2      2  33042020122    NaN   4     F   
3       10  BX410396  ...  5569    2      2  55692021222    NaN   4     F   
4       34    7468-T  ...  5569    2      2  55692021222    NaN   3     F   
..     ...       ...  ...   ...  ...    ...          ...    ...  ..   ...   
535     12   AE-2426  ...  8689    2      2  86892019122    NaN   2     F   
544     34   AP-1389  ...  9373    2      2  93732020122    NaN   7     F   
563     11   RRP6775  ...  5192    3      2  51922020132    NaN   0     F   
569      0         0  ...  6029    2      2  60292021222    NaN   4     F   
591     11    ACFG80  ...  1945    2      2  19452021222    NaN   0     F   

       cod_prod         reb   id_produto  
0           NaN  31365-ABCZ  GIRO1383626  
1           NaN   7450-GIRO  GIRO1608903  
2           NaN   3304-GIRO  GIRO1215331  
3           NaN   5569-GIRO  GIRO2022656  
4           NaN   5569-GIRO  GIRO1653303  
..          ...         ...          ...  
535         NaN   8689-GIRO  GIRO1161387  
544         NaN       52706  GIRO1518407  
563  16309606.0   5192-GIRO      RRP8637  
569         NaN  31365-ABCZ       214852  
591  11869706.0  37634-ABCZ     ISPG1286  

[459 rows x 49 columns]

# Atualizando lista de animais sem ID

In [20]:
#Tirando o sufixo da tabela
df_animais_sem_id = df_animais_sem_id.rename(columns=lambda x: x.replace('_esq', '') if '_esq' in x else x)
# Imprimir o resultado
df_animais_sem_id = pd.merge(df_animais_sem_id, df_animais_com_ID, on=['IDANIMAL'], how='left', indicator=True, suffixes=('', '_drop'))
df_animais_sem_id = df_animais_sem_id.drop(df_animais_sem_id.filter(regex='_drop$').columns, axis=1)
# Seleciona apenas as linhas que possuem valores nulos na coluna "id_produto"
df_animais_sem_id = df_animais_sem_id[df_animais_sem_id['id_produto'].isnull()]
df_animais_sem_id

reb_giro                            nomea  COMPRAC    RGVACA  \
0        7461   NANA IMPERADOR BAXTER RF SAO F     5858   6353-AH   
1       10639  LANA‡A FIV AFTERSHOCK BON.BELEM       12   9592-AG   
2        6039   EDILAINE IMPERIO FIV 4303 JARD       12  AADA0352   
3        1945             ARMA MERIDIAN DA MBL       12   1025-AN   
4       73119   CELESTE 136 TEATRO FIV COOPERR       12   1239-BF   
..        ...                              ...      ...       ...   
619      1945   JPZ QUIRERA SHAMROCK ASTER FIV       12    4032-Z   
620      2381   DALIA I FIV BRADLEY JM NOVO HO       12   4906-AZ   
621       132        QUINTA I COLE AGM TRIUNFO       34   9243-AG   
622      6915                  LIA DAS ARABIAS     5858   6125-BR   
623      6029          GERMANA FEBO FR RECREIO     5858   9069-AS   

                             nomep  GSPAI     RGPAI  \
0      IMPERADOR BAXTER VOLTA FRIA     58      1459   
1    MS ATLEES SHT AFTERSHOCK - ET     10  AX129928   
2           IMPERIO TE STA EDWIGES     11    RIG163   
3           SULLY HART MERIDIAN-ET     10  AX138576   
4               TEATRO DA SILVANIA     11    EFC383   
..                             ...    ...       ...   
619     LADYS-MANOR PL SHAMROCK-ET     10  AX135714   
620               ROCKALLI BRADLEY     10  AX117720   
621         CO-OP TOYSTORY COLE-ET     10  AX129567   
622              TURBO DAS ARABIAS   5858      0079   
623                FEBO FR RECREIO     58    6300-N   

                             nomem  GSMAE         RGMAE  ...  rac  nordc  \
0    JURITI HERBY RF SAO FRANCISCO     58        6877-N  ...    2      2   
1                 JOANE FIV KUBERA     11      ACFG1840  ...    2      2   
2    SERTEC QATAR FABIANA PROGRESS     10  HBB/B-291392  ...    2      2   
3                             FAMA     11        JNR167  ...    2      2   
4       LAGOS ASTRE QUIMICA 795 TE     10      BX335536  ...    2      2   
..                             ...    ...           ...  ...  ...    ...   
619                  JPZ ASTER FIV     11         JPZ64  ...    2      2   
620           DALIA NOVO HORIZONTE     11        JMCH76  ...    2      2   
621                   QUINTA FIV M     12        0473-O  ...    2      2   
622             LILIKA DAS ARABIAS     58       4916-AW  ...    2      2   
623            CAMPANHA FR RECREIO     58        6180-N  ...    2      2   

          rebanoe ordem  gs  sexo  cod_prod         reb  id_produto     _merge  
0     74612019122   NaN   5     F       NaN        1291         NaN  left_only  
1    106392021122   NaN   4     F       NaN  10639-GIRO         NaN  left_only  
2     60392020122   NaN   4     F       NaN         NaN         NaN  left_only  
3     19452021222   NaN   4     F       NaN  37634-ABCZ         NaN  left_only  
4    731192021222   NaN   4     F       NaN         NaN         NaN  left_only  
..            ...   ...  ..   ...       ...         ...         ...        ...  
619   19452022222   NaN   4     F       NaN  37634-ABCZ         NaN  left_only  
620   23812021222   NaN   4     F       NaN  31300-ABCZ         NaN  left_only  
621    1322019222   NaN   6     F       NaN  31835-ABCZ         NaN  left_only  
622   69152021222   NaN   5     F       NaN  27729-ABCZ         NaN  left_only  
623   60292019122   NaN   5     F       NaN  31365-ABCZ         NaN  left_only  

[583 rows x 50 columns]

# Fazendo busca no dataframe producao para encontrar id_produto dos animais sem id
conferir manualmente se entram ou não no banco

In [21]:
#Criando coluna com o ano de parto e mes parto no dataframe df_producao
df_producao['parto'] = df_producao['parto'].astype(str)
df_producao['AP'] = df_producao['parto'].str.slice(start=0, stop=4)
df_producao['MP'] = df_producao['parto'].str.slice(start=5, stop=7)
#Pegando os 10 primeiros caracteres de nome dos dataframes aniamis sem id e producao
df_producao['nome10caracteres'] = df_producao['nome'].str.slice(start=0, stop=11)
df_animais_sem_id['nome10caracteres'] = df_animais_sem_id['nomea'].str.slice(start=0, stop=11)
#removendo os espacos em branco das celulas
df_producao['nome10caracteres'] = df_producao['nome10caracteres'].str.replace('\s+', '', regex=True)
df_animais_sem_id['nome10caracteres'] = df_animais_sem_id['nome10caracteres'].str.replace('\s+', '', regex=True)

df_animais_sem_id = df_animais_sem_id.drop(df_animais_sem_id.filter(regex='_drop$').columns, axis=1)
df_producao = df_producao.drop(df_producao.filter(regex='_drop$').columns, axis=1)
df_animais_sem_id = df_animais_sem_id.drop(columns='_merge')

df_producao['AP'] = df_producao['AP'].astype(int)
df_producao['MP'] = df_producao['MP'].astype(int)
df_animais_sem_id['AP'] = df_animais_sem_id['AP'].astype(int)
df_animais_sem_id['MP'] = df_animais_sem_id['MP'].astype(int)
df_animais_com_id_na_producao = pd.merge(df_animais_sem_id, df_producao[['reb','AP', 'MP', 'nome10caracteres','id_produto']], on=['reb', 'AP', 'MP', 'nome10caracteres'], how='left', indicator=True, suffixes=('', '_drop'))
df_animais_com_id_na_producao = df_animais_com_id_na_producao[df_animais_com_id_na_producao['_merge'] == 'both']
df_animais_com_id_na_producao = df_animais_com_id_na_producao.drop(columns={'_merge', 'id_produto'})
df_animais_com_id_na_producao = df_animais_com_id_na_producao.rename(columns={'id_produto_drop': 'id_produto'})
df_animais_com_id_na_producao

reb_giro                           nomea  COMPRAC    RGVACA  \
216     11818               BACANA CABO VERDE       11  JCVL2511   
388      2381  GALENA FABULOSO JM NOVO HORIZO       38   7020-BC   
459      2381  FANTASIA FABULOSO JM NOVO HORI       38   4896-AZ   

                nomep  GSPAI   RGPAI                            nomem  GSMAE  \
216  GABEIRA GIROESTE     11  LUF182               VISCAYA CABO VERDE     11   
388  FABULOSO DO BASA     11  BASP63   ALVORADA 122 JM NOVO HORIZONTE     34   
459  FABULOSO DO BASA     11  BASP63  NOVIA‡A I FIV GOLDEN JM NOVO HO     34   

        RGMAE  ...  rac  nordc       rebanoe ordem  gs  sexo    cod_prod  \
216  JCVL1482  ...    3      2  118182021232   NaN   0     F  16530248.0   
388   2686-AG  ...    2      2   23812021222   NaN   3     F         NaN   
459    3673-U  ...    2      2   23812021222   NaN   3     F         NaN   

            reb  nome10caracteres  id_produto  
216  39977-ABCZ        BACANACABO    JCVL2511  
388  31300-ABCZ        GALENAFABU      216096  
459  31300-ABCZ        FANTASIAFA      216094  

[3 rows x 50 columns]

# Atualizando animais com id com os novos ids_produtos encontrados no dataframe de producao

In [22]:
df_animais_com_ID = pd.concat([df_animais_com_ID, df_animais_com_id_na_producao])
df_animais_com_ID

reb_giro                           nomea  COMPRAC    RGVACA  \
0        6029              DEMASIA FR RECREIO       14    0869-T   
1        7450               JACA JAYZ SJ LALU       34   6411-AH   
2        3304  CAITA OCIDENTE MJP DA CAPELA P       12    8634-X   
3        5569  LONDRINA GENGIS KHAN RECANTO D       12   0021-AW   
4        5569         BABALU METEORO DO MORRO       38   0148-AK   
..        ...                             ...      ...       ...   
569      6029                GRUTA FR RECREIO       12   4847-BG   
591      1945           ABAMA 2 FIV M. VERDE        11  ISPG1286   
216     11818               BACANA CABO VERDE       11  JCVL2511   
388      2381  GALENA FABULOSO JM NOVO HORIZO       38   7020-BC   
459      2381  FANTASIA FABULOSO JM NOVO HORI       38   4896-AZ   

                       nomep  GSPAI     RGPAI  \
0       JAGUAR TE DO GAVIAƒO     11    GAV291   
1           HORNLAND JAYZ-ET     10  AX118150   
2    OCIDENTE BEM FEITOR CAL     11   CAL4605   
3    GENGIS KHAN DE BRASILIA     11   RRP6097   
4        METEORO DE BRASILIA     11     B5226   
..                       ...    ...       ...   
569                        0      0         0   
591             C.A.SANSAƒO      11    KCA472   
216         GABEIRA GIROESTE     11    LUF182   
388         FABULOSO DO BASA     11    BASP63   
459         FABULOSO DO BASA     11    BASP63   

                               nomem  GSMAE     RGMAE  ...  rac  nordc  \
0               ERMELINDA MERGULHAO      12   AD-1407  ...    2      2   
1            GAYA FIV SANSAO SJ LALU     12    1898-P  ...    3      2   
2             MARINA DA CAPELA PAIVA     10    A-3096  ...    2      2   
3      MEARA JUDITH GOLDWYN BELLA-TE     10  BX410396  ...    2      2   
4                   SEGUNDA DO MORRO     34    7468-T  ...    2      2   
..                               ...    ...       ...  ...  ...    ...   
569                                0      0         0  ...    2      2   
591               ABAMA TE DE KUBERA     11    ACFG80  ...    2      2   
216               VISCAYA CABO VERDE     11  JCVL1482  ...    3      2   
388   ALVORADA 122 JM NOVO HORIZONTE     34   2686-AG  ...    2      2   
459  NOVIA‡A I FIV GOLDEN JM NOVO HO     34    3673-U  ...    2      2   

          rebanoe ordem  gs  sexo    cod_prod         reb   id_produto  \
0     60292021122   NaN   2     F         NaN  31365-ABCZ  GIRO1383626   
1     74502020132   NaN   6     F         NaN   7450-GIRO  GIRO1608903   
2     33042020122   NaN   4     F         NaN   3304-GIRO  GIRO1215331   
3     55692021222   NaN   4     F         NaN   5569-GIRO  GIRO2022656   
4     55692021222   NaN   3     F         NaN   5569-GIRO  GIRO1653303   
..            ...   ...  ..   ...         ...         ...          ...   
569   60292021222   NaN   4     F         NaN  31365-ABCZ       214852   
591   19452021222   NaN   0     F  11869706.0  37634-ABCZ     ISPG1286   
216  118182021232   NaN   0     F  16530248.0  39977-ABCZ     JCVL2511   
388   23812021222   NaN   3     F         NaN  31300-ABCZ       216096   
459   23812021222   NaN   3     F         NaN  31300-ABCZ       216094   

     nome10caracteres  
0                 NaN  
1                 NaN  
2                 NaN  
3                 NaN  
4                 NaN  
..                ...  
569               NaN  
591               NaN  
216        BACANACABO  
388        GALENAFABU  
459        FANTASIAFA  

[462 rows x 50 columns]

# Atualizando lista de animais sem id

In [23]:
df_animais_sem_id = pd.merge(df_animais_sem_id, df_animais_com_ID, on=['IDANIMAL'], how='left', indicator=True, suffixes=('', '_drop'))
df_animais_sem_id = df_animais_sem_id[df_animais_sem_id['id_produto_drop'].isnull()]
df_animais_sem_id = df_animais_sem_id.drop(df_animais_sem_id.filter(regex='_drop$').columns, axis=1)
# Seleciona apenas as linhas que possuem valores nulos na coluna "id_produto"
df_animais_sem_id

reb_giro                            nomea  COMPRAC    RGVACA  \
0        7461   NANA IMPERADOR BAXTER RF SAO F     5858   6353-AH   
1       10639  LANA‡A FIV AFTERSHOCK BON.BELEM       12   9592-AG   
2        6039   EDILAINE IMPERIO FIV 4303 JARD       12  AADA0352   
3        1945             ARMA MERIDIAN DA MBL       12   1025-AN   
4       73119   CELESTE 136 TEATRO FIV COOPERR       12   1239-BF   
..        ...                              ...      ...       ...   
578      1945   JPZ QUIRERA SHAMROCK ASTER FIV       12    4032-Z   
579      2381   DALIA I FIV BRADLEY JM NOVO HO       12   4906-AZ   
580       132        QUINTA I COLE AGM TRIUNFO       34   9243-AG   
581      6915                  LIA DAS ARABIAS     5858   6125-BR   
582      6029          GERMANA FEBO FR RECREIO     5858   9069-AS   

                             nomep  GSPAI     RGPAI  \
0      IMPERADOR BAXTER VOLTA FRIA     58      1459   
1    MS ATLEES SHT AFTERSHOCK - ET     10  AX129928   
2           IMPERIO TE STA EDWIGES     11    RIG163   
3           SULLY HART MERIDIAN-ET     10  AX138576   
4               TEATRO DA SILVANIA     11    EFC383   
..                             ...    ...       ...   
578     LADYS-MANOR PL SHAMROCK-ET     10  AX135714   
579               ROCKALLI BRADLEY     10  AX117720   
580         CO-OP TOYSTORY COLE-ET     10  AX129567   
581              TURBO DAS ARABIAS   5858      0079   
582                FEBO FR RECREIO     58    6300-N   

                             nomem  GSMAE         RGMAE  ...  nordc  \
0    JURITI HERBY RF SAO FRANCISCO     58        6877-N  ...      2   
1                 JOANE FIV KUBERA     11      ACFG1840  ...      2   
2    SERTEC QATAR FABIANA PROGRESS     10  HBB/B-291392  ...      2   
3                             FAMA     11        JNR167  ...      2   
4       LAGOS ASTRE QUIMICA 795 TE     10      BX335536  ...      2   
..                             ...    ...           ...  ...    ...   
578                  JPZ ASTER FIV     11         JPZ64  ...      2   
579           DALIA NOVO HORIZONTE     11        JMCH76  ...      2   
580                   QUINTA FIV M     12        0473-O  ...      2   
581             LILIKA DAS ARABIAS     58       4916-AW  ...      2   
582            CAMPANHA FR RECREIO     58        6180-N  ...      2   

          rebanoe  ordem gs  sexo  cod_prod         reb  id_produto  \
0     74612019122    NaN  5     F       NaN        1291         NaN   
1    106392021122    NaN  4     F       NaN  10639-GIRO         NaN   
2     60392020122    NaN  4     F       NaN         NaN         NaN   
3     19452021222    NaN  4     F       NaN  37634-ABCZ         NaN   
4    731192021222    NaN  4     F       NaN         NaN         NaN   
..            ...    ... ..   ...       ...         ...         ...   
578   19452022222    NaN  4     F       NaN  37634-ABCZ         NaN   
579   23812021222    NaN  4     F       NaN  31300-ABCZ         NaN   
580    1322019222    NaN  6     F       NaN  31835-ABCZ         NaN   
581   69152021222    NaN  5     F       NaN  27729-ABCZ         NaN   
582   60292019122    NaN  5     F       NaN  31365-ABCZ         NaN   

     nome10caracteres     _merge  
0          NANAIMPERA  left_only  
1           LANA‡AFIV  left_only  
2          EDILAINEIM  left_only  
3          ARMAMERIDI  left_only  
4          CELESTE136  left_only  
..                ...        ...  
578        JPZQUIRERA  left_only  
579         DALIAIFIV  left_only  
580         QUINTAICO  left_only  
581         LIADASARA  left_only  
582        GERMANAFEB  left_only  

[580 rows x 51 columns]

# Fazendo a busca na tabela genealogia_embrapa

In [24]:
#Retirando sufixo tabela genealogia
df_genealogia_embrapa = df_genealogia_embrapa.rename(columns=lambda x: x.replace('_dir', '') if '_dir' in x else x)
#Pegando ano nascimento nas colunas
df_genealogia_embrapa['nasc_est'] = df_genealogia_embrapa['nasc_est'].astype(str)
df_genealogia_embrapa['AN'] = df_genealogia_embrapa['nasc_est'].str.slice(start=0, stop=4)
df_animais_sem_id['nasc'] = df_animais_sem_id['nasc'].astype(str)
df_animais_sem_id['AN'] = df_animais_sem_id['nasc'].str.slice(start=0, stop=4)
#Criando coluna com os 10 caracteres do nome na genealogia_embrapa
df_genealogia_embrapa['nome10caracteres'] = df_genealogia_embrapa['nome'].str.slice(start=0, stop=11)
#Convertendo coluna ano e mes para tipo inteiro
#df_genealogia_embrapa['AN'] = df_genealogia_embrapa['AN'].astype(int)
#Limpando tabelas
df_animais_sem_id = df_animais_sem_id.drop(df_animais_sem_id.filter(regex='_drop$').columns, axis=1)
df_genealogia_embrapa = df_genealogia_embrapa.drop(df_genealogia_embrapa.filter(regex='_drop$').columns, axis=1)
df_animais_sem_id = df_animais_sem_id.drop(columns={'_merge'})
#criando dataframe com os ids_encontrados
df_animais_com_id_na_genealogia = pd.merge(df_animais_sem_id, df_genealogia_embrapa[['nome10caracteres','gs', 'registro_mae', 'registro_pai', 'sexo', 'AN','id_produto']], left_on=['nome10caracteres', 'gs','RGMAE', 'RGPAI', 'sexo', 'AN'], right_on=['nome10caracteres','gs', 'registro_mae', 'registro_pai', 'sexo', 'AN'], how='left', indicator=True, suffixes=('', '_drop'))
df_animais_com_id_na_genealogia = df_animais_com_id_na_genealogia[df_animais_com_id_na_genealogia['_merge'] == 'both']
df_animais_com_id_na_genealogia = df_animais_com_id_na_genealogia.drop(columns={'_merge', 'id_produto', 'registro_mae', 'registro_pai'})
df_animais_com_id_na_genealogia = df_animais_com_id_na_genealogia.rename(columns={'id_produto_drop': 'id_produto'})
df_animais_com_id_na_genealogia = df_animais_com_id_na_genealogia.drop(df_animais_com_id_na_genealogia.filter(regex='_drop$').columns, axis=1)
df_animais_com_id_na_genealogia

Empty DataFrame
Columns: [reb_giro, nomea, COMPRAC, RGVACA, nomep, GSPAI, RGPAI, nomem, GSMAE, RGMAE, DNV, MNV, ANV, UF, OP, DP, MP, AP, DURLAC, LTOT, L365, L305, DE, ME, AE, CAUSA, INTER, IDPAI, IDMAE, IDANIMAL, TORNEIO, RA, TIPO, QTD_ORDENHAS, nasc, parto, secagem, anopc, anop_menos1c, rebc, rac, nordc, rebanoe, ordem, gs, sexo, cod_prod, reb, nome10caracteres, AN, registro_mae, registro_pai, id_produto]
Index: []

[0 rows x 53 columns]

In [ ]:
#Conferência manual dos matchs com o dataframe da genealogia

In [26]:
df_animais_com_id_na_genealogia1 = pd.merge(df_animais_sem_id, df_genealogia_embrapa[['nome10caracteres','gs', 'registro_pai', 'sexo', 'AN','id_produto']], left_on=['nome10caracteres', 'gs', 'RGPAI', 'sexo', 'AN'], right_on=['nome10caracteres','gs', 'registro_pai', 'sexo', 'AN'], how='left', indicator=True, suffixes=('', '_drop'))
df_animais_com_id_na_genealogia1 = df_animais_com_id_na_genealogia1[df_animais_com_id_na_genealogia1['_merge'] == 'both']
df_animais_com_id_na_genealogia1

reb_giro     nomea  COMPRAC  RGVACA                 nomep  GSPAI   RGPAI  \
414      1945  FANTASIA       11  MBLO18  JAGUAR TE DO GAVIAƒO     11  GAV291   

                    nomem  GSMAE  RGMAE  ...  gs  sexo    cod_prod  \
414  ELECTRA FIV PEDREIRA     11  PPG41  ...   0     F  17697290.0   

            reb  id_produto  nome10caracteres    AN  registro_pai  \
414  37634-ABCZ         NaN          FANTASIA  2020        GAV291   

     id_produto_drop  _merge  
414           TAN219    both  

[1 rows x 54 columns]

# Adicionar o que ta escrito no quadro

In [27]:
df_animais_com_id_na_genealogia1 = pd.merge(df_animais_sem_id, df_genealogia_embrapa[['nome10caracteres','gs', 'sexo', 'AN','id_produto']], left_on=['nome10caracteres', 'gs', 'sexo', 'AN'], right_on=['nome10caracteres','gs', 'sexo', 'AN'], how='left', indicator=True, suffixes=('', '_drop'))
df_animais_com_id_na_genealogia1 = df_animais_com_id_na_genealogia1[df_animais_com_id_na_genealogia1['_merge'] == 'both']
df_animais_com_id_na_genealogia1

reb_giro      nomea  COMPRAC  RGVACA                    nomep  GSPAI  \
409      1945  BAILARINA       11  MBLO12  GENGIS KHAN DE BRASILIA     11   
414      1945   FANTASIA       11  MBLO18     JAGUAR TE DO GAVIAƒO     11   

       RGPAI                 nomem  GSMAE  RGMAE  ...  ordem  gs  sexo  \
409  RRP6097           BEL DA RTPA     11  NLT89  ...    NaN   0     F   
414   GAV291  ELECTRA FIV PEDREIRA     11  PPG41  ...    NaN   0     F   

       cod_prod         reb  id_produto  nome10caracteres    AN  \
409  17695912.0  37634-ABCZ         NaN         BAILARINA  2020   
414  17697290.0  37634-ABCZ         NaN          FANTASIA  2020   

     id_produto_drop  _merge  
409           RGIR34    both  
414           TAN219    both  

[2 rows x 53 columns]

In [30]:
df_animais_com_id_na_genealogia2 = pd.merge(df_animais_sem_id, df_genealogia_embrapa[['id_produto','cod_prod']], left_on=['cod_prod'], right_on=['cod_prod'], how='left', indicator=True, suffixes=('', '_drop'))
df_animais_com_id_na_genealogia2 = df_animais_com_id_na_genealogia2[df_animais_com_id_na_genealogia2['_merge'] == 'both']
df_animais_com_id_na_genealogia2

reb_giro                   nomea  COMPRAC    RGVACA  \
104      5192  URUPIANGA FIV DE BRAS.       11   RRP8772   
129      5192     PRENDA FIV DE BRAS.       11   RRP8191   
184     22541   FELICE FIV CABO VERDE       11  JCVL3334   
223     42583         BARILOCHE KENYO       11    KOK844   
269      5192      RAGIN FIV DE BRAS.       11   RRP8395   
292      5192          RUANA DE BRAS.       11   RRP8430   
304      5192              LIZ TEATRO       11    RARS10   
326     42583            MILENA KENYO       11    KOK846   
442      5192       REDE FIV DE BRAS.       11   RRP8488   
453      5192   RENUNCIA FIV DE BRAS.       11   RRP8249   
457      5192     ROTINA FIV DE BRAS.       11   RRP8273   
550      5192     PILLAR FIV DE BRAS.       11   RRP8130   
557      2389   RELIQUIA FIV DE BRAS.       11   RRP8261   

                       nomep  GSPAI    RGPAI                           nomem  \
104     JAGUAR TE DO GAVIAƒO     11   GAV291              HONDA FIV DE BRAS.   
129  GENGIS KHAN DE BRASILIA     11  RRP6097         FERVURA FIV DE BRASILIA   
184     JAGUAR TE DO GAVIAƒO     11   GAV291            ZILAH FIV CABO VERDE   
223       ELDORADO FIV KENYO     11    KOK96              IZABELLE FIV KENYO   
269     JAGUAR TE DO GAVIAƒO     11   GAV291                ISCA DE BRASILIA   
292             BACO DA RTPA     11   NLT130                   LALA DE BRAS.   
304       TEATRO DA SILVANIA     11   EFC383           RENUNCIA FIV DE BRAS.   
326            NOBRE TE CAL      11  CAL4397                JARANA FIV KENYO   
442  GENGIS KHAN DE BRASILIA     11  RRP6097                   DELI DE BRAS.   
453        FARDO FIV F.MUTUM     11   MUT697  ESFERA DE BRASILIA TN2 DE BRAS   
457        FARDO FIV F.MUTUM     11   MUT697               ILDA FIV DE BRAS.   
550  DIAMANTE TE DE BRASILIA     11  RRP5640         JAMARIA FIV DE BRASILIA   
557        FARDO FIV F.MUTUM     11   MUT697         JAMARIA FIV DE BRASILIA   

     GSMAE     RGMAE  ...  ordem  gs  sexo    cod_prod        reb  id_produto  \
104     11   RRP6268  ...    NaN   0     F  17054525.0  5192-GIRO         NaN   
129     11   RRP5998  ...    NaN   0     F  14926361.0  5192-GIRO         NaN   
184     11  JCVL1624  ...    NaN   0     F  17198554.0        NaN         NaN   
223     11    KOK526  ...    NaN   0     F  14672175.0        NaN         NaN   
269     11   RRP6699  ...    NaN   0     F  15421321.0  5192-GIRO         NaN   
292     11   RRP7119  ...    NaN   0     F  15438291.0  5192-GIRO         NaN   
304     11   RRP8249  ...    NaN   0     F  17112475.0  5192-GIRO         NaN   
326     11    KOK625  ...    NaN   0     F  14673091.0        NaN         NaN   
442     11   RRP5562  ...    NaN   0     F  15617034.0  5192-GIRO         NaN   
453     11   RRP7530  ...    NaN   0     F  15128061.0  5192-GIRO         NaN   
457     11   RRP6695  ...    NaN   0     F  15163671.0  5192-GIRO         NaN   
550     11   RRP6958  ...    NaN   0     F  14820082.0  5192-GIRO         NaN   
557     11   RRP6958  ...    NaN   0     F  15128073.0  2389-GIRO         NaN   

     nome10caracteres    AN  id_produto_drop  _merge  
104        URUPIANGAF  2019          RRP8772    both  
129         PRENDAFIV  2015          RRP8191    both  
184         FELICEFIV  2019         JCVL3334    both  
223        BARILOCHEK  2014           KOK844    both  
269         RAGINFIVD  2016          RRP8395    both  
292         RUANADEBR  2016          RRP8430    both  
304         LIZTEATRO  2019           RARS10    both  
326        MILENAKENY  2015           KOK846    both  
442         REDEFIVDE  2016          RRP8488    both  
453        RENUNCIAFI  2016          RRP8249    both  
457         ROTINAFIV  2016          RRP8273    both  
550         PILLARFIV  2015          RRP8130    both  
557        RELIQUIAFI  2016          RRP8261    both  

[13 rows x 53 columns]